<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><meta name=author content="Genomics Aotearoa"><link href=https://genomicsaotearoa.github.io/Imputation-workshop/1_tutorial/ rel=canonical><link rel=prev href=..><link href=../supplementary_1/ rel=next><link rel=icon href=../../theme_figures/nesi_ga.png><meta name=generator content="mkdocs-1.6.0, mkdocs-material-9.5.27"><title>Tutorial - Genotype Imputation</title><link rel=stylesheet href=../assets/stylesheets/main.6543a935.min.css><link rel=stylesheet href=../assets/stylesheets/palette.06af60db.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Mukta:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Mukta";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=../stylesheets/extra.css><script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script><link href=../assets/stylesheets/glightbox.min.css rel=stylesheet><style>
    html.glightbox-open { overflow: initial; height: 100%; }
    .gslide-title { margin-top: 0px; user-select: text; }
    .gslide-desc { color: #666; user-select: text; }
    .gslide-image img { background: white; }
    .gscrollbar-fixer { padding-right: 15px; }
    .gdesc-inner { font-size: 0.75rem; }
    body[data-md-color-scheme="slate"] .gdesc-inner { background: var(--md-default-bg-color);}
    body[data-md-color-scheme="slate"] .gslide-title { color: var(--md-default-fg-color);}
    body[data-md-color-scheme="slate"] .gslide-desc { color: var(--md-default-fg-color);}</style><script src=../assets/javascripts/glightbox.min.js></script></head> <body dir=ltr data-md-color-scheme=default data-md-color-primary=custom data-md-color-accent=indigo> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#tutorial class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <header class="md-header md-header--shadow" data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=.. title="Genotype Imputation" class="md-header__button md-logo" aria-label="Genotype Imputation" data-md-component=logo> <img src=../../theme_figures/nesi_ga.png alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> Genotype Imputation </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> Tutorial </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media=(prefers-color-scheme) data-md-color-scheme=default data-md-color-primary=custom data-md-color-accent=indigo aria-label="Switch to light mode" type=radio name=__palette id=__palette_0> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="m14.3 16-.7-2h-3.2l-.7 2H7.8L11 7h2l3.2 9h-1.9M20 8.69V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69m-9.15 3.96h2.3L12 9l-1.15 3.65Z"/></svg> </label> <input class=md-option data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme=default data-md-color-primary=custom data-md-color-accent=indigo aria-label="Switch to dark mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_2 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> <input class=md-option data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme=slate data-md-color-primary=custom data-md-color-accent=indigo aria-label="Switch to system preference" type=radio name=__palette id=__palette_2> <label class="md-header__button md-icon" title="Switch to system preference" for=__palette_0 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> </form> <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </label> <nav class=md-search__options aria-label=Search> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg> </button> </nav> </form> <div class=md-search__output> <div class=md-search__scrollwrap tabindex=0 data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list role=presentation></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/GenomicsAotearoa/Imputation-workshop title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> GenomicsAotearoa/Imputation-workshop </div> </a> </div> </nav> </header> <div class=md-container data-md-component=container> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=.. title="Genotype Imputation" class="md-nav__button md-logo" aria-label="Genotype Imputation" data-md-component=logo> <img src=../../theme_figures/nesi_ga.png alt=logo> </a> Genotype Imputation </label> <div class=md-nav__source> <a href=https://github.com/GenomicsAotearoa/Imputation-workshop title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> GenomicsAotearoa/Imputation-workshop </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=.. class=md-nav__link> <span class=md-ellipsis> Home </span> </a> </li> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> <span class=md-ellipsis> Tutorial </span> <span class="md-nav__icon md-icon"></span> </label> <a href=./ class="md-nav__link md-nav__link--active"> <span class=md-ellipsis> Tutorial </span> </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#procedures class=md-nav__link> <span class=md-ellipsis> Procedures </span> </a> </li> <li class=md-nav__item> <a href=#1-load-the-modules-that-are-required-on-nesi class=md-nav__link> <span class=md-ellipsis> 1. Load the modules that are required on NeSI </span> </a> </li> <li class=md-nav__item> <a href=#2-copy-the-files-into-the-home-directory class=md-nav__link> <span class=md-ellipsis> 2. Copy the files into the home directory </span> </a> </li> <li class=md-nav__item> <a href=#3-explore-the-input-file class=md-nav__link> <span class=md-ellipsis> 3. Explore the input file </span> </a> </li> <li class=md-nav__item> <a href=#4-quality-control-process class=md-nav__link> <span class=md-ellipsis> 4. Quality control process </span> </a> </li> <li class=md-nav__item> <a href=#5-generate-the-data-for-the-imputation-process class=md-nav__link> <span class=md-ellipsis> 5. Generate the data for the imputation process </span> </a> </li> <li class=md-nav__item> <a href=#6-phasing-using-beagle-5 class=md-nav__link> <span class=md-ellipsis> 6. Phasing using Beagle 5 </span> </a> </li> <li class=md-nav__item> <a href=#7-imputation-using-beagle-5 class=md-nav__link> <span class=md-ellipsis> 7. Imputation using Beagle 5 </span> </a> </li> <li class=md-nav__item> <a href=#8-imputation-using-minimac3 class=md-nav__link> <span class=md-ellipsis> 8. Imputation using minimac3 </span> </a> </li> <li class=md-nav__item> <a href=#9-calculate-the-genotype-concordance-using-vcf-compare-from-vcftools class=md-nav__link> <span class=md-ellipsis> 9. Calculate the genotype concordance using vcf-compare (from VCFtools) </span> </a> </li> <li class=md-nav__item> <a href=#10-evaluate-the-performance-of-imputation-allelicdosage-r-square class=md-nav__link> <span class=md-ellipsis> 10. Evaluate the performance of imputation: allelic/dosage R-square </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../supplementary_1/ class=md-nav__link> <span class=md-ellipsis> Supp 1: Imputation Introduction </span> </a> </li> <li class=md-nav__item> <a href=../supplementary_2/ class=md-nav__link> <span class=md-ellipsis> Supp 2: Sequencing Pipelines </span> </a> </li> <li class=md-nav__item> <a href=../supplementary_3/ class=md-nav__link> <span class=md-ellipsis> Supp 3: Tutorial </span> </a> </li> <li class=md-nav__item> <a href=../supplementary_4/ class=md-nav__link> <span class=md-ellipsis> Supp 4: Assessing Imputation Performance </span> </a> </li> <li class=md-nav__item> <a href=../supplementary_5/ class=md-nav__link> <span class=md-ellipsis> Supp 5: Stepwise Imputation up to Sequence Level and Resource Allocation </span> </a> </li> </ul> </nav> </div> </div> </div> <div class="md-sidebar md-sidebar--secondary" data-md-component=sidebar data-md-type=toc> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#procedures class=md-nav__link> <span class=md-ellipsis> Procedures </span> </a> </li> <li class=md-nav__item> <a href=#1-load-the-modules-that-are-required-on-nesi class=md-nav__link> <span class=md-ellipsis> 1. Load the modules that are required on NeSI </span> </a> </li> <li class=md-nav__item> <a href=#2-copy-the-files-into-the-home-directory class=md-nav__link> <span class=md-ellipsis> 2. Copy the files into the home directory </span> </a> </li> <li class=md-nav__item> <a href=#3-explore-the-input-file class=md-nav__link> <span class=md-ellipsis> 3. Explore the input file </span> </a> </li> <li class=md-nav__item> <a href=#4-quality-control-process class=md-nav__link> <span class=md-ellipsis> 4. Quality control process </span> </a> </li> <li class=md-nav__item> <a href=#5-generate-the-data-for-the-imputation-process class=md-nav__link> <span class=md-ellipsis> 5. Generate the data for the imputation process </span> </a> </li> <li class=md-nav__item> <a href=#6-phasing-using-beagle-5 class=md-nav__link> <span class=md-ellipsis> 6. Phasing using Beagle 5 </span> </a> </li> <li class=md-nav__item> <a href=#7-imputation-using-beagle-5 class=md-nav__link> <span class=md-ellipsis> 7. Imputation using Beagle 5 </span> </a> </li> <li class=md-nav__item> <a href=#8-imputation-using-minimac3 class=md-nav__link> <span class=md-ellipsis> 8. Imputation using minimac3 </span> </a> </li> <li class=md-nav__item> <a href=#9-calculate-the-genotype-concordance-using-vcf-compare-from-vcftools class=md-nav__link> <span class=md-ellipsis> 9. Calculate the genotype concordance using vcf-compare (from VCFtools) </span> </a> </li> <li class=md-nav__item> <a href=#10-evaluate-the-performance-of-imputation-allelicdosage-r-square class=md-nav__link> <span class=md-ellipsis> 10. Evaluate the performance of imputation: allelic/dosage R-square </span> </a> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <h1 id=tutorial>Tutorial<a class=headerlink href=#tutorial title="Permanent link">&para;</a></h1> <div class="admonition circle-info"> <p class=admonition-title>Aim</p> <ol> <li>Go through the pipeline of phasing and imputing high-density genotype to sequence level</li> <li>Understand the importance of quality control </li> <li>Know how to evaluate the imputation performance</li> </ol> </div> <div class="admonition note"> <p class=admonition-title>Import Notes</p> <ol> <li>All the parameter settings are suggestive. A different population may get absolutely different results using the same setting.</li> <li>You can't avoid experiments to find optimal settings for your own population before you started your whole genome sequence imputation. </li> <li>Beagle 5.4 and Minimac3 are shown as examples in this tutorial. You can test on other software based on your need. </li> </ol> </div> <div class="admonition screwdriver-wrench"> <p class=admonition-title>Tools we need</p> <p>BCFtools: basic bioinformatics software, in this tutorial, we use it for creating subsets and quality control. The latest version is 1.17 but we will be using 1.16. (<a href=http://samtools.github.io/bcftools/bcftools.html>http://samtools.github.io/bcftools/bcftools.html</a>)</p> <p>VCFtools: basic bioinformatics software, in this tutorial, we use it for comparing two vcf files and evaluate the concordance rate. The current version on NeSI is 0.1.15 (<a href=https://vcftools.github.io/index.html>https://vcftools.github.io/index.html</a>)</p> <p>R: basic statistics and plotting software</p> <p>Beagle: software for phasing and imputation. The current version is beagle 5.4 (version: 22Jul22.46e). (<a href=https://faculty.washington.edu/browning/beagle/beagle.html>https://faculty.washington.edu/browning/beagle/beagle.html</a>) The performance of different versions of beagle can be found here: <a href=https://www.g3journal.org/content/10/1/177>https://www.g3journal.org/content/10/1/177</a></p> <p>Minimac3: software for imputation (<a href=https://genome.sph.umich.edu/wiki/Minimac3>https://genome.sph.umich.edu/wiki/Minimac3</a>) Minimac has been updated to Minimac4 (<a href=https://github.com/statgen/Minimac4>https://github.com/statgen/Minimac4</a>) but usage requires a file format change that is done with Minimac3</p> </div> <div class="admonition file-code"> <p class=admonition-title>Input files</p> <p>In this tutorial, we used 5MB segment (30-35MB) of chr13 from the 1000 Genome phase 3, which is publicly available and can be downloaded from <a href=https://mathgen.stats.ox.ac.uk/impute/impute_v2.html>https://mathgen.stats.ox.ac.uk/impute/impute_v2.html</a></p> </div> <h2 id=procedures>Procedures<a class=headerlink href=#procedures title="Permanent link">&para;</a></h2> <h2 id=1-load-the-modules-that-are-required-on-nesi>1. Load the modules that are required on NeSI<a class=headerlink href=#1-load-the-modules-that-are-required-on-nesi title="Permanent link">&para;</a></h2> <ul> <li>To find modules that are already available on NeSI <code>module spider &lt;module_name&gt;</code></li> <li>To load modules <code>module load &lt;module_name&gt;</code> </li> </ul> <p>Open a terminal and at the prompt: </p> <div class="admonition terminal"> <p class=admonition-title>code</p> <div class=highlight><pre><span></span><code><a id=__codelineno-0-1 name=__codelineno-0-1 href=#__codelineno-0-1></a>module<span class=w> </span>purge
<a id=__codelineno-0-2 name=__codelineno-0-2 href=#__codelineno-0-2></a>module<span class=w> </span>load<span class=w> </span>BCFtools/1.16-GCC-11.3.0
<a id=__codelineno-0-3 name=__codelineno-0-3 href=#__codelineno-0-3></a>module<span class=w> </span>load<span class=w> </span>VCFtools/0.1.15-GCC-9.2.0-Perl-5.30.1
<a id=__codelineno-0-4 name=__codelineno-0-4 href=#__codelineno-0-4></a>module<span class=w> </span>load<span class=w> </span>R/4.2.1-gimkl-2022a
<a id=__codelineno-0-5 name=__codelineno-0-5 href=#__codelineno-0-5></a>module<span class=w> </span>load<span class=w> </span>Beagle/5.4-22Jul22.46e<span class=w> </span>
<a id=__codelineno-0-6 name=__codelineno-0-6 href=#__codelineno-0-6></a>module<span class=w> </span>load<span class=w> </span>Minimac3/2.0.1
</code></pre></div> </div> <h2 id=2-copy-the-files-into-the-home-directory>2. Copy the files into the home directory<a class=headerlink href=#2-copy-the-files-into-the-home-directory title="Permanent link">&para;</a></h2> <p>Define two directories: workshop directory and home directory. In this workshop, the analysis will be conducted in your <code>/home/$USER/imputation_workshop/</code> directory, (<code>~</code> is the shortcut for you home directory)</p> <div class="admonition terminal"> <p class=admonition-title>code</p> <div class=highlight><pre><span></span><code><a id=__codelineno-1-1 name=__codelineno-1-1 href=#__codelineno-1-1></a><span class=nb>cd</span><span class=w> </span>~
<a id=__codelineno-1-2 name=__codelineno-1-2 href=#__codelineno-1-2></a>mkdir<span class=w> </span>-p<span class=w> </span>imputation_workshop
<a id=__codelineno-1-3 name=__codelineno-1-3 href=#__codelineno-1-3></a><span class=nb>cd</span><span class=w> </span>~/imputation_workshop
</code></pre></div> </div> <p>The main input file (seqvcf) is extracted (from 30MB to 35MB) on chr13 from 1000 human genome data. I selected some of the reliable SNPs to generate a HD genotype dataset(hdvcf). </p> <div class="admonition terminal"> <p class=admonition-title>code</p> <div class=highlight><pre><span></span><code><a id=__codelineno-2-1 name=__codelineno-2-1 href=#__codelineno-2-1></a><span class=nv>maindir</span><span class=o>=</span>/nesi/project/nesi02659/imputation2021
<a id=__codelineno-2-2 name=__codelineno-2-2 href=#__codelineno-2-2></a><span class=nv>seqvcf</span><span class=o>=</span><span class=nv>$maindir</span>/nonfilter_seq_5MB.vcf.gz
<a id=__codelineno-2-3 name=__codelineno-2-3 href=#__codelineno-2-3></a><span class=nv>hdvcf</span><span class=o>=</span><span class=nv>$maindir</span>/hd_5MB.vcf.gz
</code></pre></div> </div> <p>Now what you need to do is to <code>cp</code> (copy) both the genotype and sequence data to your own home directory.</p> <div class="admonition terminal"> <p class=admonition-title>code</p> <div class=highlight><pre><span></span><code><a id=__codelineno-3-1 name=__codelineno-3-1 href=#__codelineno-3-1></a>cp<span class=w> </span><span class=nv>$seqvcf</span><span class=w> </span>~/imputation_workshop
<a id=__codelineno-3-2 name=__codelineno-3-2 href=#__codelineno-3-2></a>cp<span class=w> </span><span class=nv>$hdvcf</span><span class=w> </span>~/imputation_workshop
</code></pre></div> </div> <p>The genotype and sequence files use "vcf.gz" format. We can not open these compressed files directly. To check how the genotypes look, you need to use: <code>zless -S</code>. -S is to make the file well formatted and stops each line from wrapping around. </p> <div class="admonition terminal"> <p class=admonition-title>code</p> <div class=highlight><pre><span></span><code><a id=__codelineno-4-1 name=__codelineno-4-1 href=#__codelineno-4-1></a>zless<span class=w> </span>-S<span class=w> </span><span class=nv>$seqvcf</span>
<a id=__codelineno-4-2 name=__codelineno-4-2 href=#__codelineno-4-2></a>zless<span class=w> </span>-S<span class=w> </span><span class=nv>$hdvcf</span>
</code></pre></div> </div> <h2 id=3-explore-the-input-file>3. Explore the input file<a class=headerlink href=#3-explore-the-input-file title="Permanent link">&para;</a></h2> <p>When you got the dataset, the most important informaiton you may want to know is : 1. How many samples are there in my dataset 2. How many variants are there and where are they in my dataset</p> <p>To have a basic idea of the genotype, BCFtools have a very convenient function to extract this information (<a href=https://samtools.github.io/bcftools/howtos/query.html>https://samtools.github.io/bcftools/howtos/query.html</a>)</p> <div class="admonition terminal"> <p class=admonition-title>code</p> <div class=highlight><pre><span></span><code><a id=__codelineno-5-1 name=__codelineno-5-1 href=#__codelineno-5-1></a>bcftools<span class=w> </span>query<span class=w> </span>-l<span class=w> </span>nonfilter_seq_5MB.vcf.gz<span class=w> </span><span class=p>|</span><span class=w> </span>head<span class=w> </span>
<a id=__codelineno-5-2 name=__codelineno-5-2 href=#__codelineno-5-2></a>bcftools<span class=w> </span>query<span class=w> </span>-l<span class=w> </span>nonfilter_seq_5MB.vcf.gz<span class=w> </span><span class=p>|</span><span class=w> </span>wc
<a id=__codelineno-5-3 name=__codelineno-5-3 href=#__codelineno-5-3></a>
<a id=__codelineno-5-4 name=__codelineno-5-4 href=#__codelineno-5-4></a>bcftools<span class=w> </span>query<span class=w> </span>-f<span class=w> </span><span class=s1>&#39;%CHROM\t%POS\n&#39;</span><span class=w> </span>nonfilter_seq_5MB.vcf.gz<span class=w> </span><span class=p>|</span><span class=w> </span>head<span class=w> </span>
<a id=__codelineno-5-5 name=__codelineno-5-5 href=#__codelineno-5-5></a>bcftools<span class=w> </span>query<span class=w> </span>-f<span class=w> </span><span class=s1>&#39;%CHROM\t%POS\n&#39;</span><span class=w> </span>nonfilter_seq_5MB.vcf.gz<span class=w> </span><span class=p>|</span><span class=w> </span>wc<span class=w> </span>
</code></pre></div> </div> <p>If you want to have a deeper understanding of the dataset, like the number of SNPs, the number of indels, sequence depth etc, BCFtools have a very convenient function: <code>stats</code>. By checking the original sequence file's information. the code you need is as below. -s is a common tag to show "samples". Samples to include or "-" to apply all variants. Via adding this, we also generate the statistics for each individual. The output will be named: original.stats</p> <div class="admonition terminal"> <p class=admonition-title>code</p> <div class=highlight><pre><span></span><code><a id=__codelineno-6-1 name=__codelineno-6-1 href=#__codelineno-6-1></a>bcftools<span class=w> </span>stats<span class=w> </span>-s<span class=w> </span><span class=s2>&quot;-&quot;</span><span class=w> </span>nonfilter_seq_5MB.vcf.gz<span class=w> </span>&gt;<span class=w> </span>original.stats
</code></pre></div> </div> <p>Now let's have a look at the output:</p> <div class="admonition terminal"> <p class=admonition-title>code</p> <div class=highlight><pre><span></span><code><a id=__codelineno-7-1 name=__codelineno-7-1 href=#__codelineno-7-1></a>less<span class=w> </span>original.stats<span class=sb>`</span>
</code></pre></div> <details class=success> <summary>Output</summary> <p><a class=glightbox href="https://github.com/GenomicsAotearoa/Imputation-workshop/blob/master/Tutorial/img/Screen%20Shot%202020-09-16%20at%2014.37.07.png?raw=true" data-type=image data-width=auto data-height=auto data-desc-position=bottom><img alt src="https://github.com/GenomicsAotearoa/Imputation-workshop/blob/master/Tutorial/img/Screen%20Shot%202020-09-16%20at%2014.37.07.png?raw=true"></a></p> </details> </div> <p>So in the first part, you can see the basic statistics of the sequence file. We have 2504 samples in total, with 149,854 variants. 144,069 of them are SNPs, 5777 indels, 98 other, 694 multi-allelic sites with 467 multi-allelic SNPs. </p> <h2 id=4-quality-control-process>4. Quality control process<a class=headerlink href=#4-quality-control-process title="Permanent link">&para;</a></h2> <p>There are a lot of parameters you may take into consideration in your dataset, such as non-variant, singletons, multi-allelic positions, map quality, mendelian error, minor allele frequency, etc. The parameter settings vary depends highly on data, and also for the purpose of data. For example, if the data is for GWAS analysis, you probably don't want to throw away too many variants since a lot of the causals are rare alleles. If your data is used for prediction, then you probably don't want to include too much crap in the data, which is not beneficial for the next step model construction.</p> <div class="admonition terminal"> <p class=admonition-title>code</p> <div class=highlight><pre><span></span><code><a id=__codelineno-8-1 name=__codelineno-8-1 href=#__codelineno-8-1></a>bcftools<span class=w> </span>view<span class=w> </span>-O<span class=w> </span>z<span class=w> </span>-o<span class=w> </span>nosingleton.vcf.gz<span class=w> </span>-i<span class=w> </span><span class=s1>&#39;AC&gt;1&#39;</span><span class=w> </span>nonfilter_seq_5MB.vcf.gz
<a id=__codelineno-8-2 name=__codelineno-8-2 href=#__codelineno-8-2></a>tabix<span class=w> </span>-f<span class=w> </span>nosingleton.vcf.gz
<a id=__codelineno-8-3 name=__codelineno-8-3 href=#__codelineno-8-3></a>bcftools<span class=w> </span>stats<span class=w> </span>-s<span class=w> </span><span class=s2>&quot;-&quot;</span><span class=w> </span>nosingleton.vcf.gz<span class=w> </span>&gt;<span class=w> </span>step1.stats
</code></pre></div> </div> <div class="admonition terminal"> <p class=admonition-title>code</p> <div class=highlight><pre><span></span><code><a id=__codelineno-9-1 name=__codelineno-9-1 href=#__codelineno-9-1></a>less<span class=w> </span>step1.stats
</code></pre></div> <details class=success> <summary>Output</summary> <div class=highlight><pre><span></span><code><a id=__codelineno-10-1 name=__codelineno-10-1 href=#__codelineno-10-1></a><span class=c1># This file was produced by bcftools stats (1.9+htslib-1.9) and can be plotted using plot-vcfstats.</span>
<a id=__codelineno-10-2 name=__codelineno-10-2 href=#__codelineno-10-2></a><span class=c1># The command line was: bcftools stats  -s - nosingleton.vcf.gz</span>
<a id=__codelineno-10-3 name=__codelineno-10-3 href=#__codelineno-10-3></a><span class=c1>#</span>
<a id=__codelineno-10-4 name=__codelineno-10-4 href=#__codelineno-10-4></a><span class=c1># Definition of sets:</span>
<a id=__codelineno-10-5 name=__codelineno-10-5 href=#__codelineno-10-5></a><span class=c1># ID    [2]id   [3]tab-separated file names</span>
<a id=__codelineno-10-6 name=__codelineno-10-6 href=#__codelineno-10-6></a>ID<span class=w>      </span><span class=m>0</span><span class=w>       </span>nosingleton.vcf.gz
<a id=__codelineno-10-7 name=__codelineno-10-7 href=#__codelineno-10-7></a><span class=c1># SN, Summary numbers:</span>
<a id=__codelineno-10-8 name=__codelineno-10-8 href=#__codelineno-10-8></a><span class=c1>#   number of records   .. number of data rows in the VCF</span>
<a id=__codelineno-10-9 name=__codelineno-10-9 href=#__codelineno-10-9></a><span class=c1>#   number of no-ALTs   .. reference-only sites, ALT is either &quot;.&quot; or identical to REF</span>
<a id=__codelineno-10-10 name=__codelineno-10-10 href=#__codelineno-10-10></a><span class=c1>#   number of SNPs      .. number of rows with a SNP</span>
<a id=__codelineno-10-11 name=__codelineno-10-11 href=#__codelineno-10-11></a><span class=c1>#   number of MNPs      .. number of rows with a MNP, such as CC&gt;TT</span>
<a id=__codelineno-10-12 name=__codelineno-10-12 href=#__codelineno-10-12></a><span class=c1>#   number of indels    .. number of rows with an indel</span>
<a id=__codelineno-10-13 name=__codelineno-10-13 href=#__codelineno-10-13></a><span class=c1>#   number of others    .. number of rows with other type, for example a symbolic allele or</span>
<a id=__codelineno-10-14 name=__codelineno-10-14 href=#__codelineno-10-14></a><span class=c1>#                          a complex substitution, such as ACT&gt;TCGA</span>
<a id=__codelineno-10-15 name=__codelineno-10-15 href=#__codelineno-10-15></a><span class=c1>#   number of multiallelic sites     .. number of rows with multiple alternate alleles</span>
<a id=__codelineno-10-16 name=__codelineno-10-16 href=#__codelineno-10-16></a><span class=c1>#   number of multiallelic SNP sites .. number of rows with multiple alternate alleles, all SNPs</span>
<a id=__codelineno-10-17 name=__codelineno-10-17 href=#__codelineno-10-17></a><span class=c1># </span>
<a id=__codelineno-10-18 name=__codelineno-10-18 href=#__codelineno-10-18></a><span class=c1>#   Note that rows containing multiple types will be counted multiple times, in each</span>
<a id=__codelineno-10-19 name=__codelineno-10-19 href=#__codelineno-10-19></a><span class=c1>#   counter. For example, a row with a SNP and an indel increments both the SNP and</span>
<a id=__codelineno-10-20 name=__codelineno-10-20 href=#__codelineno-10-20></a><span class=c1>#   the indel counter.</span>
<a id=__codelineno-10-21 name=__codelineno-10-21 href=#__codelineno-10-21></a><span class=c1># </span>
<a id=__codelineno-10-22 name=__codelineno-10-22 href=#__codelineno-10-22></a><span class=c1># SN    [2]id   [3]key  [4]value</span>
<a id=__codelineno-10-23 name=__codelineno-10-23 href=#__codelineno-10-23></a>SN<span class=w>      </span><span class=m>0</span><span class=w>       </span>number<span class=w> </span>of<span class=w> </span>samples:<span class=w>      </span><span class=m>2504</span>
<a id=__codelineno-10-24 name=__codelineno-10-24 href=#__codelineno-10-24></a>SN<span class=w>      </span><span class=m>0</span><span class=w>       </span>number<span class=w> </span>of<span class=w> </span>records:<span class=w>      </span><span class=m>85928</span>
<a id=__codelineno-10-25 name=__codelineno-10-25 href=#__codelineno-10-25></a>SN<span class=w>      </span><span class=m>0</span><span class=w>       </span>number<span class=w> </span>of<span class=w> </span>no-ALTs:<span class=w>      </span><span class=m>0</span>
<a id=__codelineno-10-26 name=__codelineno-10-26 href=#__codelineno-10-26></a>SN<span class=w>      </span><span class=m>0</span><span class=w>       </span>number<span class=w> </span>of<span class=w> </span>SNPs:<span class=w> </span><span class=m>80184</span>
<a id=__codelineno-10-27 name=__codelineno-10-27 href=#__codelineno-10-27></a>SN<span class=w>      </span><span class=m>0</span><span class=w>       </span>number<span class=w> </span>of<span class=w> </span>MNPs:<span class=w> </span><span class=m>0</span>
<a id=__codelineno-10-28 name=__codelineno-10-28 href=#__codelineno-10-28></a>SN<span class=w>      </span><span class=m>0</span><span class=w>       </span>number<span class=w> </span>of<span class=w> </span>indels:<span class=w>       </span><span class=m>5767</span>
<a id=__codelineno-10-29 name=__codelineno-10-29 href=#__codelineno-10-29></a>SN<span class=w>      </span><span class=m>0</span><span class=w>       </span>number<span class=w> </span>of<span class=w> </span>others:<span class=w>       </span><span class=m>67</span>
<a id=__codelineno-10-30 name=__codelineno-10-30 href=#__codelineno-10-30></a>SN<span class=w>      </span><span class=m>0</span><span class=w>       </span>number<span class=w> </span>of<span class=w> </span>multiallelic<span class=w> </span>sites:<span class=w>   </span><span class=m>694</span>
<a id=__codelineno-10-31 name=__codelineno-10-31 href=#__codelineno-10-31></a>SN<span class=w>      </span><span class=m>0</span><span class=w>       </span>number<span class=w> </span>of<span class=w> </span>multiallelic<span class=w> </span>SNP<span class=w> </span>sites:<span class=w>       </span><span class=m>467</span>
<a id=__codelineno-10-32 name=__codelineno-10-32 href=#__codelineno-10-32></a><span class=c1># TSTV, transitions/transversions:</span>
<a id=__codelineno-10-33 name=__codelineno-10-33 href=#__codelineno-10-33></a><span class=c1># TSTV  [2]id   [3]ts   [4]tv   [5]ts/tv        [6]ts (1st ALT) [7]tv (1st ALT) [8]ts/tv (1st ALT)</span>
<a id=__codelineno-10-34 name=__codelineno-10-34 href=#__codelineno-10-34></a>TSTV<span class=w>    </span><span class=m>0</span><span class=w>       </span><span class=m>56037</span><span class=w>   </span><span class=m>24616</span><span class=w>   </span><span class=m>2</span>.28<span class=w>    </span><span class=m>55801</span><span class=w>   </span><span class=m>24360</span><span class=w>   </span><span class=m>2</span>.29
<a id=__codelineno-10-35 name=__codelineno-10-35 href=#__codelineno-10-35></a><span class=c1># SiS, Singleton stats:</span>
<a id=__codelineno-10-36 name=__codelineno-10-36 href=#__codelineno-10-36></a><span class=c1># SiS   [2]id   [3]allele count [4]number of SNPs       [5]number of transitions        [6]number of transversions      [7]number of indels     [8]repeat-consistent    [9]repeat-inconsistent  [10]not applicable</span>
<a id=__codelineno-10-37 name=__codelineno-10-37 href=#__codelineno-10-37></a>SiS<span class=w>     </span><span class=m>0</span><span class=w>       </span><span class=m>1</span><span class=w>       </span><span class=m>5</span><span class=w>       </span><span class=m>2</span><span class=w>       </span><span class=m>3</span><span class=w>       </span><span class=m>1</span><span class=w>       </span><span class=m>0</span><span class=w>       </span><span class=m>0</span><span class=w>       </span><span class=m>1</span>
<a id=__codelineno-10-38 name=__codelineno-10-38 href=#__codelineno-10-38></a><span class=c1># AF, Stats by non-reference allele frequency:</span>
<a id=__codelineno-10-39 name=__codelineno-10-39 href=#__codelineno-10-39></a><span class=c1># AF    [2]id   [3]allele frequency     [4]number of SNPs       [5]number of transitions        [6]number of transversions      [7]number of indels     [8]repeat-consistent    [9]repeat-inconsistent  [10]not applicable</span>
<a id=__codelineno-10-40 name=__codelineno-10-40 href=#__codelineno-10-40></a>AF<span class=w>      </span><span class=m>0</span><span class=w>       </span><span class=m>0</span>.000000<span class=w>        </span><span class=m>15765</span><span class=w>   </span><span class=m>11210</span><span class=w>   </span><span class=m>4555</span><span class=w>    </span><span class=m>30</span><span class=w>      </span><span class=m>0</span><span class=w>       </span><span class=m>0</span><span class=w>       </span><span class=m>30</span>
<a id=__codelineno-10-41 name=__codelineno-10-41 href=#__codelineno-10-41></a>AF<span class=w>      </span><span class=m>0</span><span class=w>       </span><span class=m>0</span>.000399<span class=w>        </span><span class=m>11934</span><span class=w>   </span><span class=m>8329</span><span class=w>    </span><span class=m>3605</span><span class=w>    </span><span class=m>896</span><span class=w>     </span><span class=m>0</span><span class=w>       </span><span class=m>0</span><span class=w>       </span><span class=m>896</span>
<a id=__codelineno-10-42 name=__codelineno-10-42 href=#__codelineno-10-42></a>AF<span class=w>      </span><span class=m>0</span><span class=w>       </span><span class=m>0</span>.000799<span class=w>        </span><span class=m>6258</span><span class=w>    </span><span class=m>4337</span><span class=w>    </span><span class=m>1921</span><span class=w>    </span><span class=m>453</span><span class=w>     </span><span class=m>0</span><span class=w>       </span><span class=m>0</span><span class=w>       </span><span class=m>453</span>
<a id=__codelineno-10-43 name=__codelineno-10-43 href=#__codelineno-10-43></a>AF<span class=w>      </span><span class=m>0</span><span class=w>       </span><span class=m>0</span>.001198<span class=w>        </span><span class=m>4171</span><span class=w>    </span><span class=m>2866</span><span class=w>    </span><span class=m>1305</span><span class=w>    </span><span class=m>299</span><span class=w>     </span><span class=m>0</span><span class=w>       </span><span class=m>0</span><span class=w>       </span><span class=m>299</span>
<a id=__codelineno-10-44 name=__codelineno-10-44 href=#__codelineno-10-44></a>AF<span class=w>      </span><span class=m>0</span><span class=w>       </span><span class=m>0</span>.001597<span class=w>        </span><span class=m>2834</span><span class=w>    </span><span class=m>1940</span><span class=w>    </span><span class=m>894</span><span class=w>     </span><span class=m>209</span><span class=w>     </span><span class=m>0</span><span class=w>       </span><span class=m>0</span><span class=w>       </span><span class=m>209</span>
<a id=__codelineno-10-45 name=__codelineno-10-45 href=#__codelineno-10-45></a>AF<span class=w>      </span><span class=m>0</span><span class=w>       </span><span class=m>0</span>.001997<span class=w>        </span><span class=m>2371</span><span class=w>    </span><span class=m>1664</span><span class=w>    </span><span class=m>707</span><span class=w>     </span><span class=m>172</span><span class=w>     </span><span class=m>0</span><span class=w>       </span><span class=m>0</span><span class=w>       </span><span class=m>172</span>
<a id=__codelineno-10-46 name=__codelineno-10-46 href=#__codelineno-10-46></a>AF<span class=w>      </span><span class=m>0</span><span class=w>       </span><span class=m>0</span>.002396<span class=w>        </span><span class=m>1851</span><span class=w>    </span><span class=m>1295</span><span class=w>    </span><span class=m>556</span><span class=w>     </span><span class=m>123</span><span class=w>     </span><span class=m>0</span><span class=w>       </span><span class=m>0</span><span class=w>       </span><span class=m>123</span>
<a id=__codelineno-10-47 name=__codelineno-10-47 href=#__codelineno-10-47></a>AF<span class=w>      </span><span class=m>0</span><span class=w>       </span><span class=m>0</span>.002796<span class=w>        </span><span class=m>1588</span><span class=w>    </span><span class=m>1120</span><span class=w>    </span><span class=m>468</span><span class=w>     </span><span class=m>111</span><span class=w>     </span><span class=m>0</span><span class=w>       </span><span class=m>0</span><span class=w>       </span><span class=m>111</span>
<a id=__codelineno-10-48 name=__codelineno-10-48 href=#__codelineno-10-48></a>AF<span class=w>      </span><span class=m>0</span><span class=w>       </span><span class=m>0</span>.003195<span class=w>        </span><span class=m>1380</span><span class=w>    </span><span class=m>957</span><span class=w>     </span><span class=m>423</span><span class=w>     </span><span class=m>114</span><span class=w>     </span><span class=m>0</span><span class=w>       </span><span class=m>0</span><span class=w>       </span><span class=m>114</span>
<a id=__codelineno-10-49 name=__codelineno-10-49 href=#__codelineno-10-49></a>AF<span class=w>      </span><span class=m>0</span><span class=w>       </span><span class=m>0</span>.003594<span class=w>        </span><span class=m>1160</span><span class=w>    </span><span class=m>780</span><span class=w>     </span><span class=m>380</span><span class=w>     </span><span class=m>95</span><span class=w>      </span><span class=m>0</span><span class=w>       </span><span class=m>0</span><span class=w>       </span><span class=m>95</span>
<a id=__codelineno-10-50 name=__codelineno-10-50 href=#__codelineno-10-50></a>AF<span class=w>      </span><span class=m>0</span><span class=w>       </span><span class=m>0</span>.003994<span class=w>        </span><span class=m>988</span><span class=w>     </span><span class=m>688</span><span class=w>     </span><span class=m>300</span><span class=w>     </span><span class=m>68</span><span class=w>      </span><span class=m>0</span><span class=w>       </span><span class=m>0</span><span class=w>       </span><span class=m>68</span>
<a id=__codelineno-10-51 name=__codelineno-10-51 href=#__codelineno-10-51></a>AF<span class=w>      </span><span class=m>0</span><span class=w>       </span><span class=m>0</span>.004393<span class=w>        </span><span class=m>923</span><span class=w>     </span><span class=m>633</span><span class=w>     </span><span class=m>290</span><span class=w>     </span><span class=m>77</span><span class=w>      </span><span class=m>0</span><span class=w>       </span><span class=m>0</span><span class=w>       </span><span class=m>77</span>
</code></pre></div> </details> </div> <p>As we can see here, after removing the non-variants and singletons, the number of variants decreased to 85,928, 80,184 SNPs, 5767 indels, 67 others, 694 multiallelic sites with 467 multi-allelic SNPs. </p> <div class="admonition terminal"> <p class=admonition-title>code</p> <div class=highlight><pre><span></span><code><a id=__codelineno-11-1 name=__codelineno-11-1 href=#__codelineno-11-1></a>bcftools<span class=w> </span>view<span class=w> </span>-O<span class=w> </span>z<span class=w> </span>-o<span class=w> </span>nosingleton_2alleles.vcf.gz<span class=w> </span>--max-alleles<span class=w> </span><span class=m>2</span><span class=w> </span>nosingleton.vcf.gz
<a id=__codelineno-11-2 name=__codelineno-11-2 href=#__codelineno-11-2></a>tabix<span class=w> </span>-f<span class=w> </span>nosingleton_2alleles.vcf.gz
<a id=__codelineno-11-3 name=__codelineno-11-3 href=#__codelineno-11-3></a>bcftools<span class=w> </span>stats<span class=w> </span>-s<span class=w> </span><span class=s2>&quot;-&quot;</span><span class=w> </span>nosingleton_2alleles.vcf.gz<span class=w> </span>&gt;<span class=w> </span>step2.stats
</code></pre></div> </div> <div class="admonition terminal"> <p class=admonition-title>code</p> <div class=highlight><pre><span></span><code><a id=__codelineno-12-1 name=__codelineno-12-1 href=#__codelineno-12-1></a>less<span class=w> </span>step2.stats
</code></pre></div> <details class=success> <summary>Output</summary> <div class=highlight><pre><span></span><code><a id=__codelineno-13-1 name=__codelineno-13-1 href=#__codelineno-13-1></a><span class=c1># This file was produced by bcftools stats (1.9+htslib-1.9) and can be plotted using plot-vcfstats.</span>
<a id=__codelineno-13-2 name=__codelineno-13-2 href=#__codelineno-13-2></a><span class=c1># The command line was: bcftools stats  -s - nosingleton_2alleles.vcf.gz</span>
<a id=__codelineno-13-3 name=__codelineno-13-3 href=#__codelineno-13-3></a><span class=c1>#</span>
<a id=__codelineno-13-4 name=__codelineno-13-4 href=#__codelineno-13-4></a><span class=c1># Definition of sets:</span>
<a id=__codelineno-13-5 name=__codelineno-13-5 href=#__codelineno-13-5></a><span class=c1># ID    [2]id   [3]tab-separated file names</span>
<a id=__codelineno-13-6 name=__codelineno-13-6 href=#__codelineno-13-6></a>ID<span class=w>      </span><span class=m>0</span><span class=w>       </span>nosingleton_2alleles.vcf.gz
<a id=__codelineno-13-7 name=__codelineno-13-7 href=#__codelineno-13-7></a><span class=c1># SN, Summary numbers:</span>
<a id=__codelineno-13-8 name=__codelineno-13-8 href=#__codelineno-13-8></a><span class=c1>#   number of records   .. number of data rows in the VCF</span>
<a id=__codelineno-13-9 name=__codelineno-13-9 href=#__codelineno-13-9></a><span class=c1>#   number of no-ALTs   .. reference-only sites, ALT is either &quot;.&quot; or identical to REF</span>
<a id=__codelineno-13-10 name=__codelineno-13-10 href=#__codelineno-13-10></a><span class=c1>#   number of SNPs      .. number of rows with a SNP</span>
<a id=__codelineno-13-11 name=__codelineno-13-11 href=#__codelineno-13-11></a><span class=c1>#   number of MNPs      .. number of rows with a MNP, such as CC&gt;TT</span>
<a id=__codelineno-13-12 name=__codelineno-13-12 href=#__codelineno-13-12></a><span class=c1>#   number of indels    .. number of rows with an indel</span>
<a id=__codelineno-13-13 name=__codelineno-13-13 href=#__codelineno-13-13></a><span class=c1>#   number of others    .. number of rows with other type, for example a symbolic allele or</span>
<a id=__codelineno-13-14 name=__codelineno-13-14 href=#__codelineno-13-14></a><span class=c1>#                          a complex substitution, such as ACT&gt;TCGA</span>
<a id=__codelineno-13-15 name=__codelineno-13-15 href=#__codelineno-13-15></a><span class=c1>#   number of multiallelic sites     .. number of rows with multiple alternate alleles</span>
<a id=__codelineno-13-16 name=__codelineno-13-16 href=#__codelineno-13-16></a><span class=c1>#   number of multiallelic SNP sites .. number of rows with multiple alternate alleles, all SNPs</span>
<a id=__codelineno-13-17 name=__codelineno-13-17 href=#__codelineno-13-17></a><span class=c1># </span>
<a id=__codelineno-13-18 name=__codelineno-13-18 href=#__codelineno-13-18></a><span class=c1>#   Note that rows containing multiple types will be counted multiple times, in each</span>
<a id=__codelineno-13-19 name=__codelineno-13-19 href=#__codelineno-13-19></a><span class=c1>#   counter. For example, a row with a SNP and an indel increments both the SNP and</span>
<a id=__codelineno-13-20 name=__codelineno-13-20 href=#__codelineno-13-20></a><span class=c1>#   the indel counter.</span>
<a id=__codelineno-13-21 name=__codelineno-13-21 href=#__codelineno-13-21></a><span class=c1># </span>
<a id=__codelineno-13-22 name=__codelineno-13-22 href=#__codelineno-13-22></a><span class=c1># SN    [2]id   [3]key  [4]value</span>
<a id=__codelineno-13-23 name=__codelineno-13-23 href=#__codelineno-13-23></a>SN<span class=w>      </span><span class=m>0</span><span class=w>       </span>number<span class=w> </span>of<span class=w> </span>samples:<span class=w>      </span><span class=m>2504</span>
<a id=__codelineno-13-24 name=__codelineno-13-24 href=#__codelineno-13-24></a>SN<span class=w>      </span><span class=m>0</span><span class=w>       </span>number<span class=w> </span>of<span class=w> </span>records:<span class=w>      </span><span class=m>85234</span>
<a id=__codelineno-13-25 name=__codelineno-13-25 href=#__codelineno-13-25></a>SN<span class=w>      </span><span class=m>0</span><span class=w>       </span>number<span class=w> </span>of<span class=w> </span>no-ALTs:<span class=w>      </span><span class=m>0</span>
<a id=__codelineno-13-26 name=__codelineno-13-26 href=#__codelineno-13-26></a>SN<span class=w>      </span><span class=m>0</span><span class=w>       </span>number<span class=w> </span>of<span class=w> </span>SNPs:<span class=w> </span><span class=m>79627</span>
<a id=__codelineno-13-27 name=__codelineno-13-27 href=#__codelineno-13-27></a>SN<span class=w>      </span><span class=m>0</span><span class=w>       </span>number<span class=w> </span>of<span class=w> </span>MNPs:<span class=w> </span><span class=m>0</span>
<a id=__codelineno-13-28 name=__codelineno-13-28 href=#__codelineno-13-28></a>SN<span class=w>      </span><span class=m>0</span><span class=w>       </span>number<span class=w> </span>of<span class=w> </span>indels:<span class=w>       </span><span class=m>5543</span>
<a id=__codelineno-13-29 name=__codelineno-13-29 href=#__codelineno-13-29></a>SN<span class=w>      </span><span class=m>0</span><span class=w>       </span>number<span class=w> </span>of<span class=w> </span>others:<span class=w>       </span><span class=m>64</span>
<a id=__codelineno-13-30 name=__codelineno-13-30 href=#__codelineno-13-30></a>SN<span class=w>      </span><span class=m>0</span><span class=w>       </span>number<span class=w> </span>of<span class=w> </span>multiallelic<span class=w> </span>sites:<span class=w>   </span><span class=m>0</span>
<a id=__codelineno-13-31 name=__codelineno-13-31 href=#__codelineno-13-31></a>SN<span class=w>      </span><span class=m>0</span><span class=w>       </span>number<span class=w> </span>of<span class=w> </span>multiallelic<span class=w> </span>SNP<span class=w> </span>sites:<span class=w>       </span><span class=m>0</span>
<a id=__codelineno-13-32 name=__codelineno-13-32 href=#__codelineno-13-32></a><span class=c1># TSTV, transitions/transversions:</span>
<a id=__codelineno-13-33 name=__codelineno-13-33 href=#__codelineno-13-33></a><span class=c1># TSTV  [2]id   [3]ts   [4]tv   [5]ts/tv        [6]ts (1st ALT) [7]tv (1st ALT) [8]ts/tv (1st ALT)</span>
<a id=__codelineno-13-34 name=__codelineno-13-34 href=#__codelineno-13-34></a>TSTV<span class=w>    </span><span class=m>0</span><span class=w>       </span><span class=m>55586</span><span class=w>   </span><span class=m>24041</span><span class=w>   </span><span class=m>2</span>.31<span class=w>    </span><span class=m>55586</span><span class=w>   </span><span class=m>24041</span><span class=w>   </span><span class=m>2</span>.31
<a id=__codelineno-13-35 name=__codelineno-13-35 href=#__codelineno-13-35></a><span class=c1># SiS, Singleton stats:</span>
<a id=__codelineno-13-36 name=__codelineno-13-36 href=#__codelineno-13-36></a><span class=c1># SiS   [2]id   [3]allele count [4]number of SNPs       [5]number of transitions        [6]number of transversions      [7]number of indels     [8]repeat-consistent    [9]repeat-inconsistent  [10]not applicable</span>
<a id=__codelineno-13-37 name=__codelineno-13-37 href=#__codelineno-13-37></a>SiS<span class=w>     </span><span class=m>0</span><span class=w>       </span><span class=m>1</span><span class=w>       </span><span class=m>0</span><span class=w>       </span><span class=m>0</span><span class=w>       </span><span class=m>0</span><span class=w>       </span><span class=m>0</span><span class=w>       </span><span class=m>0</span><span class=w>       </span><span class=m>0</span><span class=w>       </span><span class=m>0</span>
<a id=__codelineno-13-38 name=__codelineno-13-38 href=#__codelineno-13-38></a><span class=c1># AF, Stats by non-reference allele frequency:</span>
<a id=__codelineno-13-39 name=__codelineno-13-39 href=#__codelineno-13-39></a><span class=c1># AF    [2]id   [3]allele frequency     [4]number of SNPs       [5]number of transitions        [6]number of transversions      [7]number of indels     [8]repeat-consistent    [9]repeat-inconsistent  [10]not applicable</span>
<a id=__codelineno-13-40 name=__codelineno-13-40 href=#__codelineno-13-40></a>AF<span class=w>      </span><span class=m>0</span><span class=w>       </span><span class=m>0</span>.000000<span class=w>        </span><span class=m>15598</span><span class=w>   </span><span class=m>11131</span><span class=w>   </span><span class=m>4467</span><span class=w>    </span><span class=m>29</span><span class=w>      </span><span class=m>0</span><span class=w>       </span><span class=m>0</span><span class=w>       </span><span class=m>29</span>
<a id=__codelineno-13-41 name=__codelineno-13-41 href=#__codelineno-13-41></a>AF<span class=w>      </span><span class=m>0</span><span class=w>       </span><span class=m>0</span>.000399<span class=w>        </span><span class=m>11775</span><span class=w>   </span><span class=m>8248</span><span class=w>    </span><span class=m>3527</span><span class=w>    </span><span class=m>857</span><span class=w>     </span><span class=m>0</span><span class=w>       </span><span class=m>0</span><span class=w>       </span><span class=m>857</span>
<a id=__codelineno-13-42 name=__codelineno-13-42 href=#__codelineno-13-42></a>AF<span class=w>      </span><span class=m>0</span><span class=w>       </span><span class=m>0</span>.000799<span class=w>        </span><span class=m>6163</span><span class=w>    </span><span class=m>4293</span><span class=w>    </span><span class=m>1870</span><span class=w>    </span><span class=m>433</span><span class=w>     </span><span class=m>0</span><span class=w>       </span><span class=m>0</span><span class=w>       </span><span class=m>433</span>
<a id=__codelineno-13-43 name=__codelineno-13-43 href=#__codelineno-13-43></a>AF<span class=w>      </span><span class=m>0</span><span class=w>       </span><span class=m>0</span>.001198<span class=w>        </span><span class=m>4104</span><span class=w>    </span><span class=m>2833</span><span class=w>    </span><span class=m>1271</span><span class=w>    </span><span class=m>292</span><span class=w>     </span><span class=m>0</span><span class=w>       </span><span class=m>0</span><span class=w>       </span><span class=m>292</span>
<a id=__codelineno-13-44 name=__codelineno-13-44 href=#__codelineno-13-44></a>AF<span class=w>      </span><span class=m>0</span><span class=w>       </span><span class=m>0</span>.001597<span class=w>        </span><span class=m>2806</span><span class=w>    </span><span class=m>1928</span><span class=w>    </span><span class=m>878</span><span class=w>     </span><span class=m>202</span><span class=w>     </span><span class=m>0</span><span class=w>       </span><span class=m>0</span><span class=w>       </span><span class=m>202</span>
<a id=__codelineno-13-45 name=__codelineno-13-45 href=#__codelineno-13-45></a>AF<span class=w>      </span><span class=m>0</span><span class=w>       </span><span class=m>0</span>.001997<span class=w>        </span><span class=m>2344</span><span class=w>    </span><span class=m>1651</span><span class=w>    </span><span class=m>693</span><span class=w>     </span><span class=m>166</span><span class=w>     </span><span class=m>0</span><span class=w>       </span><span class=m>0</span><span class=w>       </span><span class=m>166</span>
<a id=__codelineno-13-46 name=__codelineno-13-46 href=#__codelineno-13-46></a>AF<span class=w>      </span><span class=m>0</span><span class=w>       </span><span class=m>0</span>.002396<span class=w>        </span><span class=m>1830</span><span class=w>    </span><span class=m>1286</span><span class=w>    </span><span class=m>544</span><span class=w>     </span><span class=m>118</span><span class=w>     </span><span class=m>0</span><span class=w>       </span><span class=m>0</span><span class=w>       </span><span class=m>118</span>
<a id=__codelineno-13-47 name=__codelineno-13-47 href=#__codelineno-13-47></a>AF<span class=w>      </span><span class=m>0</span><span class=w>       </span><span class=m>0</span>.002796<span class=w>        </span><span class=m>1567</span><span class=w>    </span><span class=m>1111</span><span class=w>    </span><span class=m>456</span><span class=w>     </span><span class=m>106</span><span class=w>     </span><span class=m>0</span><span class=w>       </span><span class=m>0</span><span class=w>       </span><span class=m>106</span>
<a id=__codelineno-13-48 name=__codelineno-13-48 href=#__codelineno-13-48></a>AF<span class=w>      </span><span class=m>0</span><span class=w>       </span><span class=m>0</span>.003195<span class=w>        </span><span class=m>1358</span><span class=w>    </span><span class=m>951</span><span class=w>     </span><span class=m>407</span><span class=w>     </span><span class=m>106</span><span class=w>     </span><span class=m>0</span><span class=w>       </span><span class=m>0</span><span class=w>       </span><span class=m>106</span>
<a id=__codelineno-13-49 name=__codelineno-13-49 href=#__codelineno-13-49></a>AF<span class=w>      </span><span class=m>0</span><span class=w>       </span><span class=m>0</span>.003594<span class=w>        </span><span class=m>1144</span><span class=w>    </span><span class=m>775</span><span class=w>     </span><span class=m>369</span><span class=w>     </span><span class=m>85</span><span class=w>      </span><span class=m>0</span><span class=w>       </span><span class=m>0</span><span class=w>       </span><span class=m>85</span>
<a id=__codelineno-13-50 name=__codelineno-13-50 href=#__codelineno-13-50></a>AF<span class=w>      </span><span class=m>0</span><span class=w>       </span><span class=m>0</span>.003994<span class=w>        </span><span class=m>972</span><span class=w>     </span><span class=m>682</span><span class=w>     </span><span class=m>290</span><span class=w>     </span><span class=m>64</span><span class=w>      </span><span class=m>0</span><span class=w>       </span><span class=m>0</span><span class=w>       </span><span class=m>64</span>
<a id=__codelineno-13-51 name=__codelineno-13-51 href=#__codelineno-13-51></a>AF<span class=w>      </span><span class=m>0</span><span class=w>       </span><span class=m>0</span>.004393<span class=w>        </span><span class=m>914</span><span class=w>     </span><span class=m>629</span><span class=w>     </span><span class=m>285</span><span class=w>     </span><span class=m>71</span><span class=w>      </span><span class=m>0</span><span class=w>       </span><span class=m>0</span><span class=w>       </span><span class=m>71</span>
</code></pre></div> </details> </div> <p>As we can see here, after setting the maximum allele into 2, the multi-allelic variants should be gone. The total number of variants decreased to 85,234, 79,627 SNPs, 5543 indels and 64 others. In this tutorial, I am only gonna consider non-variant, singleton and multi-allelic variants since the majority of the imputation software can not handle them anyway. Other filtering processes can also be done using bcftools, you can just pop in the website and go to filtering sessions. </p> <h2 id=5-generate-the-data-for-the-imputation-process>5. Generate the data for the imputation process<a class=headerlink href=#5-generate-the-data-for-the-imputation-process title="Permanent link">&para;</a></h2> <p>We are gonna just use this data for imputation. So there are two populations involved: reference population and study population. In reality you should have both datasets ready. In this tutorial, I decide to just use this one dataset. Treat the first 1000 samples as my reference, and the rest will be set as my study population. Using bcftools to extract the sample ID and basic awk function to generate two population ID files as follow:</p> <div class="admonition terminal"> <p class=admonition-title>code</p> <div class=highlight><pre><span></span><code><a id=__codelineno-14-1 name=__codelineno-14-1 href=#__codelineno-14-1></a>bcftools<span class=w> </span>query<span class=w> </span>-l<span class=w> </span>nonfilter_seq_5MB.vcf.gz<span class=w> </span>&gt;<span class=w> </span>seq.ID
<a id=__codelineno-14-2 name=__codelineno-14-2 href=#__codelineno-14-2></a>awk<span class=w> </span><span class=s1>&#39;NR&gt;0&amp;&amp;NR&lt;=1000&#39;</span><span class=w> </span>seq.ID<span class=w> </span>&gt;<span class=w> </span>ref.ID
<a id=__codelineno-14-3 name=__codelineno-14-3 href=#__codelineno-14-3></a>awk<span class=w> </span><span class=s1>&#39;NR&gt;1000&amp;&amp;NR&lt;=2000&#39;</span><span class=w> </span>seq.ID<span class=w> </span>&gt;<span class=w> </span>study.ID
</code></pre></div> </div> <p>Now we create a new directory for our imputation:</p> <div class="admonition terminal"> <p class=admonition-title>code</p> <div class=highlight><pre><span></span><code><a id=__codelineno-15-1 name=__codelineno-15-1 href=#__codelineno-15-1></a><span class=c1>#confirm the current working directory is ~/imputation_workshop` as imputations directory will be created inside the former</span>
<a id=__codelineno-15-2 name=__codelineno-15-2 href=#__codelineno-15-2></a><span class=nb>pwd</span>
</code></pre></div> <details class=success> <summary>Output</summary> <div class=highlight><pre><span></span><code><a id=__codelineno-16-1 name=__codelineno-16-1 href=#__codelineno-16-1></a>/home/&lt;userid&gt;/imputation_workshop
</code></pre></div> </details> <div class=highlight><pre><span></span><code><a id=__codelineno-17-1 name=__codelineno-17-1 href=#__codelineno-17-1></a>mkdir<span class=w> </span>-p<span class=w> </span>imputation
<a id=__codelineno-17-2 name=__codelineno-17-2 href=#__codelineno-17-2></a><span class=nb>cd</span><span class=w> </span>imputation
</code></pre></div> </div> <p>bcftools is very handy for extracting the sample from the whole dataset via using -S. We will generate two files for our study samples: one from HD genotype as my study population. And another one from filtered sequence data for future validation. After extraction, we use the function <code>tabix</code> to index both files.</p> <div class="admonition terminal"> <p class=admonition-title>code</p> <div class=highlight><pre><span></span><code><a id=__codelineno-18-1 name=__codelineno-18-1 href=#__codelineno-18-1></a>bcftools<span class=w> </span>view<span class=w> </span>-O<span class=w> </span>z<span class=w> </span>-o<span class=w> </span>study_hd.vcf.gz<span class=w> </span>-S<span class=w> </span>~/imputation_workshop/study.ID<span class=w> </span>~/imputation_workshop/hd_5MB.vcf.gz
<a id=__codelineno-18-2 name=__codelineno-18-2 href=#__codelineno-18-2></a>tabix<span class=w> </span>-f<span class=w> </span>study_hd.vcf.gz
<a id=__codelineno-18-3 name=__codelineno-18-3 href=#__codelineno-18-3></a>bcftools<span class=w> </span>view<span class=w> </span>-O<span class=w> </span>z<span class=w> </span>-o<span class=w> </span>study_filtered.vcf.gz<span class=w> </span>-S<span class=w> </span>~/imputation_workshop/study.ID<span class=w> </span>~/imputation_workshop/nosingleton_2alleles.vcf.gz<span class=w> </span><span class=c1>#for validation</span>
<a id=__codelineno-18-4 name=__codelineno-18-4 href=#__codelineno-18-4></a>tabix<span class=w> </span>-f<span class=w> </span>study_filtered.vcf.gz
</code></pre></div> </div> <p>Similarly, we will extract the samples for our reference samples. The same functions will be used here. We will generate two files for the reference samples, one is from the original unfiltered sequence, and another one from the filtered sequence. Again, we will use tabix to index these two files. </p> <div class="admonition terminal"> <p class=admonition-title>code</p> <div class=highlight><pre><span></span><code><a id=__codelineno-19-1 name=__codelineno-19-1 href=#__codelineno-19-1></a>bcftools<span class=w> </span>view<span class=w> </span>-O<span class=w> </span>z<span class=w> </span>-o<span class=w> </span>ref_nonfiltered.vcf.gz<span class=w> </span>-S<span class=w> </span>~/imputation_workshop/ref.ID<span class=w> </span>~/imputation_workshop/nonfilter_seq_5MB.vcf.gz
<a id=__codelineno-19-2 name=__codelineno-19-2 href=#__codelineno-19-2></a>tabix<span class=w> </span>-f<span class=w> </span>ref_nonfiltered.vcf.gz
<a id=__codelineno-19-3 name=__codelineno-19-3 href=#__codelineno-19-3></a>bcftools<span class=w> </span>view<span class=w> </span>-O<span class=w> </span>z<span class=w> </span>-o<span class=w> </span>ref_filtered.vcf.gz<span class=w> </span>-S<span class=w> </span>~/imputation_workshop/ref.ID<span class=w> </span>~/imputation_workshop/nosingleton_2alleles.vcf.gz
<a id=__codelineno-19-4 name=__codelineno-19-4 href=#__codelineno-19-4></a>tabix<span class=w> </span>-f<span class=w> </span>ref_filtered.vcf.gz
</code></pre></div> </div> <h2 id=6-phasing-using-beagle-5>6. Phasing using Beagle 5<a class=headerlink href=#6-phasing-using-beagle-5 title="Permanent link">&para;</a></h2> <p>For imputation, no matter which software are you using, phasing is compulsory for the reference population. For some software like minimac3, both study and reference population have to be phased. If you simply pop the unphased reference population into your imputation software, the software will immediately give you an error message. </p> <p>The data I downloaded already finished phasing that you can see in the dataset, all the genotypes are phased (eg: "0|1", not "0/1"). Besides phasing needs a lot of computation resources and a certain amount of time. Here since the data is already phased, it won't take too long. </p> <div class="admonition terminal"> <p class=admonition-title>code</p> <div class=highlight><pre><span></span><code><a id=__codelineno-20-1 name=__codelineno-20-1 href=#__codelineno-20-1></a>beagle<span class=w> </span><span class=nv>gt</span><span class=o>=</span>ref_filtered.vcf.gz<span class=w> </span><span class=nv>chrom</span><span class=o>=</span><span class=m>13</span><span class=w> </span><span class=nv>out</span><span class=o>=</span>ref_filtered_phased
</code></pre></div> <details class=success> <summary>Output</summary> <div class=highlight><pre><span></span><code><a id=__codelineno-21-1 name=__codelineno-21-1 href=#__codelineno-21-1></a>beagle.22Jul22.46e.jar<span class=w> </span><span class=o>(</span>version<span class=w> </span><span class=m>5</span>.4<span class=o>)</span>
<a id=__codelineno-21-2 name=__codelineno-21-2 href=#__codelineno-21-2></a>Copyright<span class=w> </span><span class=o>(</span>C<span class=o>)</span><span class=w> </span><span class=m>2014</span>-2022<span class=w> </span>Brian<span class=w> </span>L.<span class=w> </span>Browning
<a id=__codelineno-21-3 name=__codelineno-21-3 href=#__codelineno-21-3></a>Enter<span class=w> </span><span class=s2>&quot;java -jar beagle.22Jul22.46e.jar&quot;</span><span class=w> </span>to<span class=w> </span>list<span class=w> </span><span class=nb>command</span><span class=w> </span>line<span class=w> </span>argument
<a id=__codelineno-21-4 name=__codelineno-21-4 href=#__codelineno-21-4></a>Start<span class=w> </span>time:<span class=w> </span><span class=m>09</span>:05<span class=w> </span>AM<span class=w> </span>NZST<span class=w> </span>on<span class=w> </span><span class=m>20</span><span class=w> </span>Jun<span class=w> </span><span class=m>2023</span>
<a id=__codelineno-21-5 name=__codelineno-21-5 href=#__codelineno-21-5></a>
<a id=__codelineno-21-6 name=__codelineno-21-6 href=#__codelineno-21-6></a>Command<span class=w> </span>line:<span class=w> </span>java<span class=w> </span>-Xmx27305m<span class=w> </span>-jar<span class=w> </span>beagle.22Jul22.46e.jar
<a id=__codelineno-21-7 name=__codelineno-21-7 href=#__codelineno-21-7></a><span class=w>  </span><span class=nv>gt</span><span class=o>=</span>ref_filtered.vcf.gz
<a id=__codelineno-21-8 name=__codelineno-21-8 href=#__codelineno-21-8></a><span class=w>  </span><span class=nv>chrom</span><span class=o>=</span><span class=m>13</span>
<a id=__codelineno-21-9 name=__codelineno-21-9 href=#__codelineno-21-9></a><span class=w>  </span><span class=nv>out</span><span class=o>=</span>ref_filtered_phased
<a id=__codelineno-21-10 name=__codelineno-21-10 href=#__codelineno-21-10></a><span class=w>  </span><span class=nv>nthreads</span><span class=o>=</span><span class=m>4</span>
<a id=__codelineno-21-11 name=__codelineno-21-11 href=#__codelineno-21-11></a>
<a id=__codelineno-21-12 name=__codelineno-21-12 href=#__codelineno-21-12></a>No<span class=w> </span>genetic<span class=w> </span>map<span class=w> </span>is<span class=w> </span>specified:<span class=w> </span>using<span class=w> </span><span class=m>1</span><span class=w> </span><span class=nv>cM</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>1</span><span class=w> </span>Mb
<a id=__codelineno-21-13 name=__codelineno-21-13 href=#__codelineno-21-13></a>
<a id=__codelineno-21-14 name=__codelineno-21-14 href=#__codelineno-21-14></a>Reference<span class=w> </span>samples:<span class=w>                    </span><span class=m>0</span>
<a id=__codelineno-21-15 name=__codelineno-21-15 href=#__codelineno-21-15></a>Study<span class=w>     </span>samples:<span class=w>                </span><span class=m>1</span>,000
<a id=__codelineno-21-16 name=__codelineno-21-16 href=#__codelineno-21-16></a>
<a id=__codelineno-21-17 name=__codelineno-21-17 href=#__codelineno-21-17></a>Window<span class=w> </span><span class=m>1</span><span class=w> </span><span class=o>[</span><span class=m>13</span>:30000193-34999935<span class=o>]</span>
<a id=__codelineno-21-18 name=__codelineno-21-18 href=#__codelineno-21-18></a>Study<span class=w>     </span>markers:<span class=w>               </span><span class=m>85</span>,234
<a id=__codelineno-21-19 name=__codelineno-21-19 href=#__codelineno-21-19></a>
<a id=__codelineno-21-20 name=__codelineno-21-20 href=#__codelineno-21-20></a>Cumulative<span class=w> </span>Statistics:
<a id=__codelineno-21-21 name=__codelineno-21-21 href=#__codelineno-21-21></a>
<a id=__codelineno-21-22 name=__codelineno-21-22 href=#__codelineno-21-22></a>Study<span class=w>     </span>markers:<span class=w>               </span><span class=m>85</span>,234
<a id=__codelineno-21-23 name=__codelineno-21-23 href=#__codelineno-21-23></a>
<a id=__codelineno-21-24 name=__codelineno-21-24 href=#__codelineno-21-24></a>Total<span class=w> </span>time:<span class=w>                    </span><span class=m>16</span><span class=w> </span>seconds
<a id=__codelineno-21-25 name=__codelineno-21-25 href=#__codelineno-21-25></a>
<a id=__codelineno-21-26 name=__codelineno-21-26 href=#__codelineno-21-26></a>End<span class=w> </span>time:<span class=w> </span><span class=m>09</span>:05<span class=w> </span>AM<span class=w> </span>NZST<span class=w> </span>on<span class=w> </span><span class=m>20</span><span class=w> </span>June<span class=w> </span><span class=m>2021</span>
</code></pre></div> </details> </div> <div class="admonition terminal"> <p class=admonition-title>code</p> <div class=highlight><pre><span></span><code><a id=__codelineno-22-1 name=__codelineno-22-1 href=#__codelineno-22-1></a>tabix<span class=w> </span>-f<span class=w> </span>ref_filtered_phased.vcf.gz
</code></pre></div> </div> <div class="admonition terminal"> <p class=admonition-title>code</p> <div class=highlight><pre><span></span><code><a id=__codelineno-23-1 name=__codelineno-23-1 href=#__codelineno-23-1></a>beagle<span class=w> </span><span class=nv>gt</span><span class=o>=</span>ref_nonfiltered.vcf.gz<span class=w> </span><span class=nv>chrom</span><span class=o>=</span><span class=m>13</span><span class=w> </span><span class=nv>out</span><span class=o>=</span>ref_nonfiltered_phased
<a id=__codelineno-23-2 name=__codelineno-23-2 href=#__codelineno-23-2></a>
<a id=__codelineno-23-3 name=__codelineno-23-3 href=#__codelineno-23-3></a>tabix<span class=w> </span>-f<span class=w> </span>ref_nonfiltered_phased.vcf.gz
</code></pre></div> </div> <h2 id=7-imputation-using-beagle-5>7. Imputation using Beagle 5<a class=headerlink href=#7-imputation-using-beagle-5 title="Permanent link">&para;</a></h2> <p>In this tutorial, I will show you the imputation using two software: Beagle 5 and minimac3. Both software are very stable, reliable, easy-to-use, free, and pretty popular. Beagle 5 is computationally demanding but can give you accurate results very fast. Minimac is computationally efficient, but a bit slower. In addition, to prove that quality control is an important procedure, both filtered reference and unfiltered sequence reference will be used.</p> <p>Beagle has been evolved from version 3.0 to the current 5.4 version. It has become much faster and simpler. And able to handle large datasets. In the meantime, the parameters for running the software have been reduced a lot. There are several important parameters that can influence imputation performance such as effective population size (Ne), window size, etc. Check the following paper: Improving Imputation Quality in BEAGLE for Crop and Livestock Data <a href=https://www.g3journal.org/content/10/1/177>https://www.g3journal.org/content/10/1/177</a></p> <div class="admonition terminal"> <p class=admonition-title>code</p> <div class=highlight><pre><span></span><code><a id=__codelineno-24-1 name=__codelineno-24-1 href=#__codelineno-24-1></a>beagle<span class=w> </span><span class=nv>gt</span><span class=o>=</span>study_hd.vcf.gz<span class=w> </span><span class=nv>ref</span><span class=o>=</span>ref_filtered_phased.vcf.gz<span class=w> </span><span class=nv>chrom</span><span class=o>=</span><span class=m>13</span><span class=w> </span><span class=nv>impute</span><span class=o>=</span><span class=nb>true</span><span class=w> </span><span class=nv>gp</span><span class=o>=</span><span class=nb>true</span><span class=w> </span><span class=nv>out</span><span class=o>=</span>HD_to_seq_filtered_beagle5
<a id=__codelineno-24-2 name=__codelineno-24-2 href=#__codelineno-24-2></a>tabix<span class=w> </span>-f<span class=w> </span>HD_to_seq_filtered_beagle5.vcf.gz
</code></pre></div> </div> <div class="admonition terminal"> <p class=admonition-title>code</p> <div class=highlight><pre><span></span><code><a id=__codelineno-25-1 name=__codelineno-25-1 href=#__codelineno-25-1></a>beagle<span class=w> </span><span class=nv>gt</span><span class=o>=</span>study_hd.vcf.gz<span class=w> </span><span class=nv>ref</span><span class=o>=</span>ref_nonfiltered_phased.vcf.gz<span class=w> </span><span class=nv>chrom</span><span class=o>=</span><span class=m>13</span><span class=w> </span><span class=nv>impute</span><span class=o>=</span><span class=nb>true</span><span class=w> </span><span class=nv>gp</span><span class=o>=</span><span class=nb>true</span><span class=w> </span><span class=nv>out</span><span class=o>=</span>HD_to_seq_nonfiltered_beagle5
<a id=__codelineno-25-2 name=__codelineno-25-2 href=#__codelineno-25-2></a>tabix<span class=w> </span>-f<span class=w> </span>HD_to_seq_nonfiltered_beagle5.vcf.gz
</code></pre></div> </div> <h2 id=8-imputation-using-minimac3>8. Imputation using minimac3<a class=headerlink href=#8-imputation-using-minimac3 title="Permanent link">&para;</a></h2> <p>The imputation process for using minimac3 is rather similar. It is more efficient than Beagle 5 but slightly slower. It takes ~10 mins to impute to filtered sequence reference and ~15 mins to impute to unfiltered sequence reference. I have already done the process, so you can just copy the outputs from the project folder to the current imputation folder. </p> <div class="admonition terminal"> <p class=admonition-title>code</p> <div class=highlight><pre><span></span><code><a id=__codelineno-26-1 name=__codelineno-26-1 href=#__codelineno-26-1></a><span class=c1>#Minimac3 --refHaps ref_filtered_phased.vcf.gz --haps study_hd.vcf.gz --prefix HD_to_seq_filtered_minimac3</span>
<a id=__codelineno-26-2 name=__codelineno-26-2 href=#__codelineno-26-2></a><span class=c1>#tabix -f HD_to_seq_filtered_minimac3.dose.vcf.gz</span>
<a id=__codelineno-26-3 name=__codelineno-26-3 href=#__codelineno-26-3></a><span class=c1>#Minimac3 --refHaps ref_nonfiltered_phased.vcf.gz --haps study_hd.vcf.gz --prefix HD_to_seq_nonfiltered_minimac3</span>
<a id=__codelineno-26-4 name=__codelineno-26-4 href=#__codelineno-26-4></a><span class=c1>#tabix -f HD_to_seq_nonfiltered_minimac3.dose.vcf.gz</span>
</code></pre></div> </div> <div class="admonition terminal"> <p class=admonition-title>code</p> <div class=highlight><pre><span></span><code><a id=__codelineno-27-1 name=__codelineno-27-1 href=#__codelineno-27-1></a>cp<span class=w> </span><span class=nv>$maindir</span>/imputation/HD_to_seq_filtered_minimac3.*<span class=w> </span>~/imputation_workshop/imputation
<a id=__codelineno-27-2 name=__codelineno-27-2 href=#__codelineno-27-2></a>cp<span class=w> </span><span class=nv>$maindir</span>/imputation/HD_to_seq_nonfiltered_minimac3.*<span class=w> </span>~/imputation_workshop/imputation
</code></pre></div> </div> <h2 id=9-calculate-the-genotype-concordance-using-vcf-compare-from-vcftools>9. Calculate the genotype concordance using vcf-compare (from VCFtools)<a class=headerlink href=#9-calculate-the-genotype-concordance-using-vcf-compare-from-vcftools title="Permanent link">&para;</a></h2> <p>In this tutorial, I am going show you two parameters: genotype concordance and allelic/dosage R-square.</p> <p>To compare two vcfs and have an idea of genotype concordance, there is a sub-function from vcftools: vcf-compare. so just pop in <code>vcf-compare VCF1 VCF2 &gt; output</code>. You will have an output file.</p> <p>In the previous session, we have four imputation outputs using both Beagle 5 and minimac3, to impute to filtered and unfiltered sequence reference. So four concordance file will be generated as below:</p> <div class="admonition terminal"> <p class=admonition-title>code</p> <div class=highlight><pre><span></span><code><a id=__codelineno-28-1 name=__codelineno-28-1 href=#__codelineno-28-1></a>vcf-compare<span class=w> </span>study_filtered.vcf.gz<span class=w> </span>HD_to_seq_filtered_beagle5.vcf.gz<span class=w> </span>&gt;<span class=w> </span>concordance_beagle5_filtered
<a id=__codelineno-28-2 name=__codelineno-28-2 href=#__codelineno-28-2></a>less<span class=w> </span>concordance_beagle5_filtered
</code></pre></div> </div> <p><a class=glightbox href="https://github.com/GenomicsAotearoa/Imputation-workshop/blob/master/Tutorial/img/Screen%20Shot%202020-09-16%20at%2014.50.01.png?raw=true" data-type=image data-width=auto data-height=auto data-desc-position=bottom><img alt src="https://github.com/GenomicsAotearoa/Imputation-workshop/blob/master/Tutorial/img/Screen%20Shot%202020-09-16%20at%2014.50.01.png?raw=true"></a></p> <div class="admonition terminal"> <p class=admonition-title>code</p> <div class=highlight><pre><span></span><code><a id=__codelineno-29-1 name=__codelineno-29-1 href=#__codelineno-29-1></a>vcf-compare<span class=w> </span>study_filtered.vcf.gz<span class=w> </span>HD_to_seq_nonfiltered_beagle5.vcf.gz<span class=w> </span>&gt;<span class=w> </span>concordance_beagle5_nonfiltered
<a id=__codelineno-29-2 name=__codelineno-29-2 href=#__codelineno-29-2></a>less<span class=w> </span>concordance_beagle5_nonfiltered
</code></pre></div> </div> <p><a class=glightbox href="https://github.com/GenomicsAotearoa/Imputation-workshop/blob/master/Tutorial/img/Screen%20Shot%202020-09-16%20at%2014.51.21.png?raw=true" data-type=image data-width=auto data-height=auto data-desc-position=bottom><img alt src="https://github.com/GenomicsAotearoa/Imputation-workshop/blob/master/Tutorial/img/Screen%20Shot%202020-09-16%20at%2014.51.21.png?raw=true"></a></p> <div class="admonition terminal"> <p class=admonition-title>code</p> <div class=highlight><pre><span></span><code><a id=__codelineno-30-1 name=__codelineno-30-1 href=#__codelineno-30-1></a>vcf-compare<span class=w> </span>study_filtered.vcf.gz<span class=w> </span>HD_to_seq_filtered_minimac3.dose.vcf.gz<span class=w> </span>&gt;<span class=w> </span>concordance_minimac3_filtered
<a id=__codelineno-30-2 name=__codelineno-30-2 href=#__codelineno-30-2></a>less<span class=w> </span>concordance_minimac3_filtered
</code></pre></div> </div> <p><a class=glightbox href="https://github.com/GenomicsAotearoa/Imputation-workshop/blob/master/Tutorial/img/Screen%20Shot%202020-09-16%20at%2014.52.43.png?raw=true" data-type=image data-width=auto data-height=auto data-desc-position=bottom><img alt src="https://github.com/GenomicsAotearoa/Imputation-workshop/blob/master/Tutorial/img/Screen%20Shot%202020-09-16%20at%2014.52.43.png?raw=true"></a></p> <div class="admonition terminal"> <p class=admonition-title>code</p> <div class=highlight><pre><span></span><code><a id=__codelineno-31-1 name=__codelineno-31-1 href=#__codelineno-31-1></a>vcf-compare<span class=w> </span>study_filtered.vcf.gz<span class=w> </span>HD_to_seq_nonfiltered_minimac3.dose.vcf.gz<span class=w> </span>&gt;<span class=w> </span>concordance_minimac3_nonfiltered
<a id=__codelineno-31-2 name=__codelineno-31-2 href=#__codelineno-31-2></a>less<span class=w> </span>concordance_minimac3_nonfiltered
</code></pre></div> </div> <p><a class=glightbox href="https://github.com/GenomicsAotearoa/Imputation-workshop/blob/master/Tutorial/img/Screen%20Shot%202020-09-16%20at%2014.53.47.png?raw=true" data-type=image data-width=auto data-height=auto data-desc-position=bottom><img alt src="https://github.com/GenomicsAotearoa/Imputation-workshop/blob/master/Tutorial/img/Screen%20Shot%202020-09-16%20at%2014.53.47.png?raw=true"></a></p> <h2 id=10-evaluate-the-performance-of-imputation-allelicdosage-r-square>10. Evaluate the performance of imputation: allelic/dosage R-square<a class=headerlink href=#10-evaluate-the-performance-of-imputation-allelicdosage-r-square title="Permanent link">&para;</a></h2> <p>To calculate the dosage R-square, beagle 5 does not provide a seperate file. You may need a bit code to extract the information:</p> <div class="admonition terminal"> <p class=admonition-title>code</p> <div class=highlight><pre><span></span><code><a id=__codelineno-32-1 name=__codelineno-32-1 href=#__codelineno-32-1></a>bcftools<span class=w> </span>query<span class=w> </span>-f<span class=w> </span><span class=s1>&#39;%CHROM\t%POS\t%ID\t%REF\t%ALT\t%QUAL\t%FILTER\t%DR2\t%AF\t%IMP\n&#39;</span><span class=w> </span>HD_to_seq_filtered_beagle5.vcf.gz<span class=w> </span>&gt;<span class=w> </span>HD_to_seq_filtered_beagle5.r2
<a id=__codelineno-32-2 name=__codelineno-32-2 href=#__codelineno-32-2></a>bcftools<span class=w> </span>query<span class=w> </span>-f<span class=w> </span><span class=s1>&#39;%CHROM\t%POS\t%ID\t%REF\t%ALT\t%QUAL\t%FILTER\t%DR2\t%AF\t%IMP\n&#39;</span><span class=w> </span>HD_to_seq_nonfiltered_beagle5.vcf.gz<span class=w> </span>&gt;<span class=w> </span>HD_to_seq_nonfiltered_beagle5.r2
</code></pre></div> </div> <p>The columns of the file we generated are chromosome, position, SNP name, reference allele, alternative allele, quality, filter, <strong>dosage r-square,</strong> allele frequency, whether it is imputed. It is a thoughtful enough file that provides us all the information, the only additional part we may need to do is calculate the minor allele frequency from allele frequency.</p> <div class="admonition r-project"> <p class=admonition-title>code</p> <div class=highlight><pre><span></span><code><a id=__codelineno-33-1 name=__codelineno-33-1 href=#__codelineno-33-1></a>head<span class=w> </span>HD_to_seq_filtered_beagle5.r2
</code></pre></div> </div> <p><a class=glightbox href="https://github.com/GenomicsAotearoa/Imputation-workshop/blob/master/Tutorial/img/Screen%20Shot%202020-09-16%20at%2014.56.35.png?raw=true" data-type=image data-width=auto data-height=auto data-desc-position=bottom><img alt src="https://github.com/GenomicsAotearoa/Imputation-workshop/blob/master/Tutorial/img/Screen%20Shot%202020-09-16%20at%2014.56.35.png?raw=true"></a></p> <p>minimac3 generates an info file after it finishes imputing. It is pretty thoughtful that it provides us the minor allele frequency directly. The troubling part is that we have to extract the position from the SNP column for future comparison. </p> <div class="admonition terminal"> <p class=admonition-title>code</p> <div class=highlight><pre><span></span><code><a id=__codelineno-34-1 name=__codelineno-34-1 href=#__codelineno-34-1></a>head<span class=w> </span>HD_to_seq_filtered_minimac3.info
</code></pre></div> </div> <p><a class=glightbox href="https://github.com/GenomicsAotearoa/Imputation-workshop/blob/master/Tutorial/img/Screen%20Shot%202020-09-16%20at%2015.02.07.png?raw=true" data-type=image data-width=auto data-height=auto data-desc-position=bottom><img alt src="https://github.com/GenomicsAotearoa/Imputation-workshop/blob/master/Tutorial/img/Screen%20Shot%202020-09-16%20at%2015.02.07.png?raw=true"></a></p> <p>So we have four output here, and we are going pop them in R to have a look: </p> <p><code>R</code></p> <p>We either use R at the terminal, or you can use the RStudio associated with Jupyter. When plotting at a terminal, you will need to write the output to a file and open the image from the Jupyter file directory on the left.</p> <p>The first step is to read in all our output files in R</p> <div class="admonition r-project"> <p class=admonition-title>code</p> <div class=highlight><pre><span></span><code><a id=__codelineno-35-1 name=__codelineno-35-1 href=#__codelineno-35-1></a><span class=nf>setwd</span><span class=p>(</span><span class=s>&quot;~/imputation_workshop/imputation&quot;</span><span class=p>)</span>
<a id=__codelineno-35-2 name=__codelineno-35-2 href=#__codelineno-35-2></a><span class=n>filteredBG5</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>read.table</span><span class=p>(</span><span class=s>&quot;HD_to_seq_filtered_beagle5.r2&quot;</span><span class=p>)</span>
<a id=__codelineno-35-3 name=__codelineno-35-3 href=#__codelineno-35-3></a><span class=n>nonfilteredBG5</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>read.table</span><span class=p>(</span><span class=s>&quot;HD_to_seq_nonfiltered_beagle5.r2&quot;</span><span class=p>)</span>
<a id=__codelineno-35-4 name=__codelineno-35-4 href=#__codelineno-35-4></a><span class=n>filteredminimac3</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>read.table</span><span class=p>(</span><span class=s>&quot;HD_to_seq_filtered_minimac3.info&quot;</span><span class=p>,</span><span class=w> </span><span class=n>header</span><span class=o>=</span><span class=bp>T</span><span class=p>)</span>
<a id=__codelineno-35-5 name=__codelineno-35-5 href=#__codelineno-35-5></a><span class=n>nonfilteredminimac3</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>read.table</span><span class=p>(</span><span class=s>&quot;HD_to_seq_nonfiltered_minimac3.info&quot;</span><span class=p>,</span><span class=w> </span><span class=n>header</span><span class=o>=</span><span class=bp>T</span><span class=p>)</span>
</code></pre></div> </div> <p>Then we need to extract all the positions. This step is a bit redundant for beagle outputs but really helpful for the minimac3 output. The function we are gonna use is <code>substr</code>, it tells R to just extract the string from the 4<sup>th</sup> digit to the 11<sup>th</sup> digit. </p> <div class="admonition r-project"> <p class=admonition-title>code</p> <div class=highlight><pre><span></span><code><a id=__codelineno-36-1 name=__codelineno-36-1 href=#__codelineno-36-1></a><span class=n>filteredBG5</span><span class=o>$</span><span class=n>Pos</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>filteredBG5</span><span class=o>$</span><span class=n>V2</span>
<a id=__codelineno-36-2 name=__codelineno-36-2 href=#__codelineno-36-2></a><span class=n>nonfilteredBG5</span><span class=o>$</span><span class=n>Pos</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>nonfilteredBG5</span><span class=o>$</span><span class=n>V2</span>
<a id=__codelineno-36-3 name=__codelineno-36-3 href=#__codelineno-36-3></a><span class=n>filteredminimac3</span><span class=o>$</span><span class=n>Pos</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>substr</span><span class=p>(</span><span class=n>filteredminimac3</span><span class=o>$</span><span class=n>SNP</span><span class=p>,</span><span class=w> </span><span class=m>4</span><span class=p>,</span><span class=w> </span><span class=m>11</span><span class=p>)</span>
<a id=__codelineno-36-4 name=__codelineno-36-4 href=#__codelineno-36-4></a><span class=n>nonfilteredminimac3</span><span class=o>$</span><span class=n>Pos</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>substr</span><span class=p>(</span><span class=n>nonfilteredminimac3</span><span class=o>$</span><span class=n>SNP</span><span class=p>,</span><span class=w> </span><span class=m>4</span><span class=p>,</span><span class=w> </span><span class=m>11</span><span class=p>)</span>
</code></pre></div> </div> <p>The next step is to extract the R-square for beagle 5. Usually, it shouldn't be a problem, you get the number in that column directly. However, in this session, we used the unfiltered reference, which contains the multi-allelic positions. In this case, Beagle will give you multiple possible solutions for those multi-allelic positions. In this case, we just take the first solution to make things easier. Here you may see a warning message mentioned <code>NA</code> generated. Don't worry about that. </p> <div class="admonition r-project"> <p class=admonition-title>code</p> <div class=highlight><pre><span></span><code><a id=__codelineno-37-1 name=__codelineno-37-1 href=#__codelineno-37-1></a><span class=n>filteredBG5</span><span class=o>$</span><span class=n>DR2_filtered_BG5</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>filteredBG5</span><span class=o>$</span><span class=n>V8</span>
<a id=__codelineno-37-2 name=__codelineno-37-2 href=#__codelineno-37-2></a><span class=n>nonfilteredBG5</span><span class=o>$</span><span class=n>DR2_nonfiltered_BG5</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>as.numeric</span><span class=p>(</span><span class=nf>substr</span><span class=p>(</span><span class=n>nonfilteredBG5</span><span class=o>$</span><span class=n>V8</span><span class=p>,</span><span class=w> </span><span class=m>1</span><span class=p>,</span><span class=w> </span><span class=m>4</span><span class=p>))</span>
<a id=__codelineno-37-3 name=__codelineno-37-3 href=#__codelineno-37-3></a><span class=n>filteredminimac3</span><span class=o>$</span><span class=n>Rsq_filtered_minimac3</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>filteredminimac3</span><span class=o>$</span><span class=n>Rsq</span>
<a id=__codelineno-37-4 name=__codelineno-37-4 href=#__codelineno-37-4></a><span class=n>nonfilteredminimac3</span><span class=o>$</span><span class=n>Rsq_nonfiltered_minimac3</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>nonfilteredminimac3</span><span class=o>$</span><span class=n>Rsq</span>
</code></pre></div> </div> <p>Now let's merge both the output files from Beagle and Minimac3, then final merge them into a file called <code>finalmerge</code></p> <div class="admonition r-project"> <p class=admonition-title>code</p> <div class=highlight><pre><span></span><code><a id=__codelineno-38-1 name=__codelineno-38-1 href=#__codelineno-38-1></a><span class=n>mergedBeagle</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>merge</span><span class=p>(</span><span class=n>filteredBG5</span><span class=p>,</span><span class=w> </span><span class=n>nonfilteredBG5</span><span class=p>,</span><span class=w> </span><span class=n>by.x</span><span class=o>=</span><span class=s>&quot;Pos&quot;</span><span class=p>,</span><span class=w> </span><span class=n>by.y</span><span class=o>=</span><span class=s>&quot;Pos&quot;</span><span class=p>,</span><span class=w> </span><span class=n>all</span><span class=o>=</span><span class=kc>FALSE</span><span class=p>)</span>
<a id=__codelineno-38-2 name=__codelineno-38-2 href=#__codelineno-38-2></a><span class=n>mergedMinimac3</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>merge</span><span class=p>(</span><span class=n>filteredminimac3</span><span class=p>,</span><span class=w> </span><span class=n>nonfilteredminimac3</span><span class=p>,</span><span class=w> </span><span class=n>by.x</span><span class=o>=</span><span class=s>&quot;Pos&quot;</span><span class=p>,</span><span class=w> </span><span class=n>by.y</span><span class=o>=</span><span class=s>&quot;Pos&quot;</span><span class=p>,</span><span class=w> </span><span class=n>all</span><span class=o>=</span><span class=kc>FALSE</span><span class=p>)</span>
<a id=__codelineno-38-3 name=__codelineno-38-3 href=#__codelineno-38-3></a><span class=n>finalmerge</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>merge</span><span class=p>(</span><span class=n>mergedBeagle</span><span class=p>,</span><span class=w> </span><span class=n>mergedMinimac3</span><span class=p>,</span><span class=w> </span><span class=n>by.x</span><span class=o>=</span><span class=s>&quot;Pos&quot;</span><span class=p>,</span><span class=w> </span><span class=n>by.y</span><span class=o>=</span><span class=s>&quot;Pos&quot;</span><span class=p>,</span><span class=w> </span><span class=n>all</span><span class=o>=</span><span class=kc>FALSE</span><span class=p>)</span>
</code></pre></div> </div> <p>Let's have a look at the summary</p> <div class="admonition r-project"> <p class=admonition-title>code</p> <div class=highlight><pre><span></span><code><a id=__codelineno-39-1 name=__codelineno-39-1 href=#__codelineno-39-1></a><span class=nf>summary</span><span class=p>(</span><span class=n>finalmerge</span><span class=o>$</span><span class=n>DR2_filtered_BG5</span><span class=p>)</span>
<a id=__codelineno-39-2 name=__codelineno-39-2 href=#__codelineno-39-2></a><span class=nf>summary</span><span class=p>(</span><span class=n>finalmerge</span><span class=o>$</span><span class=n>DR2_nonfiltered_BG5</span><span class=p>)</span>
<a id=__codelineno-39-3 name=__codelineno-39-3 href=#__codelineno-39-3></a><span class=nf>summary</span><span class=p>(</span><span class=n>finalmerge</span><span class=o>$</span><span class=n>Rsq_filtered_minimac3</span><span class=p>)</span>
<a id=__codelineno-39-4 name=__codelineno-39-4 href=#__codelineno-39-4></a><span class=nf>summary</span><span class=p>(</span><span class=n>finalmerge</span><span class=o>$</span><span class=n>Rsq_nonfiltered_minimac3</span><span class=p>)</span>
</code></pre></div> </div> <p>In both cases of beagle 5 and minimac3, using unfiltered reference gave us poorer performance compared to the filtered ones. Minimac3 gave slightly higher allelic square compare to beagle5. It is not always the case since we are only using 5MB here. And also the performance depends on a lot of parameters. As I mentioned, Beagle is fast but computationally demanding. Minimac 3 is slower but very efficient. There are of course other software for you to choose. Which software to use, what parameters for QC, questions such as those I may not have an answer, you have to figure it out by doing experiments. </p> <p>As I also mentioned allelic/dosage-r square is a good parameter for evaluating the performance. Here you can see the relationship between MAF and allelic/dosage-r square. </p> <p><a class=glightbox href="https://github.com/GenomicsAotearoa/Imputation-workshop/blob/master/Tutorial/img/Screen%20Shot%202020-09-16%20at%2015.13.24.png?raw=true" data-type=image data-width=auto data-height=auto data-desc-position=bottom><img alt src="https://github.com/GenomicsAotearoa/Imputation-workshop/blob/master/Tutorial/img/Screen%20Shot%202020-09-16%20at%2015.13.24.png?raw=true"></a></p> <div class="admonition r-project"> <p class=admonition-title>code</p> <div class=highlight><pre><span></span><code><a id=__codelineno-40-1 name=__codelineno-40-1 href=#__codelineno-40-1></a><span class=nf>plot</span><span class=p>(</span><span class=n>finalmerge</span><span class=o>$</span><span class=n>MAF.x</span><span class=p>,</span><span class=n>finalmerge</span><span class=o>$</span><span class=n>DR2_filtered_BG5</span><span class=p>,</span><span class=w> </span><span class=n>pch</span><span class=o>=</span><span class=m>4</span><span class=p>)</span>
</code></pre></div> </div> <p><a class=glightbox href="https://github.com/GenomicsAotearoa/Imputation-workshop/blob/master/Tutorial/img/image003.png?raw=true" data-type=image data-width=auto data-height=auto data-desc-position=bottom><img alt src="https://github.com/GenomicsAotearoa/Imputation-workshop/blob/master/Tutorial/img/image003.png?raw=true"></a></p> <div class="admonition r-project"> <p class=admonition-title>code</p> <div class=highlight><pre><span></span><code><a id=__codelineno-41-1 name=__codelineno-41-1 href=#__codelineno-41-1></a><span class=nf>plot</span><span class=p>(</span><span class=n>finalmerge</span><span class=o>$</span><span class=n>MAF.x</span><span class=p>,</span><span class=n>finalmerge</span><span class=o>$</span><span class=n>Rsq_filtered_minimac3</span><span class=p>,</span><span class=w> </span><span class=n>pch</span><span class=o>=</span><span class=m>4</span><span class=p>)</span>
</code></pre></div> </div> <p><a class=glightbox href="https://github.com/GenomicsAotearoa/Imputation-workshop/blob/master/Tutorial/img/image002.png?raw=true" data-type=image data-width=auto data-height=auto data-desc-position=bottom><img alt src="https://github.com/GenomicsAotearoa/Imputation-workshop/blob/master/Tutorial/img/image002.png?raw=true"></a></p> <p>And also, we can also have a look at the correlation between the allelic/dosage-r square from beagle 5 and minimac3. </p> <div class="admonition r-project"> <p class=admonition-title>code</p> <div class=highlight><pre><span></span><code><a id=__codelineno-42-1 name=__codelineno-42-1 href=#__codelineno-42-1></a><span class=nf>plot</span><span class=p>(</span><span class=n>finalmerge</span><span class=o>$</span><span class=n>DR2_filtered_BG5</span><span class=p>,</span><span class=n>finalmerge</span><span class=o>$</span><span class=n>Rsq_filtered_minimac3</span><span class=p>,</span><span class=w> </span><span class=n>pch</span><span class=o>=</span><span class=m>4</span><span class=p>)</span>
</code></pre></div> </div> <p><a class=glightbox href="https://github.com/GenomicsAotearoa/Imputation-workshop/blob/master/Tutorial/img/image001.png?raw=true" data-type=image data-width=auto data-height=auto data-desc-position=bottom><img alt src="https://github.com/GenomicsAotearoa/Imputation-workshop/blob/master/Tutorial/img/image001.png?raw=true"></a></p> <p>Now let's have a look at the poorly imputed regions, the simple way will be just use the position as X-axis and accuracy as Y-axis. The command will be just </p> <div class="admonition r-project"> <p class=admonition-title>code</p> <div class=highlight><pre><span></span><code><a id=__codelineno-43-1 name=__codelineno-43-1 href=#__codelineno-43-1></a><span class=nf>plot</span><span class=p>(</span><span class=n>finalmerge</span><span class=o>$</span><span class=n>Pos</span><span class=p>,</span><span class=n>finalmerge</span><span class=o>$</span><span class=n>DR2_filtered_BG5</span><span class=p>)</span>
<a id=__codelineno-43-2 name=__codelineno-43-2 href=#__codelineno-43-2></a><span class=nf>plot</span><span class=p>(</span><span class=n>finalmerge</span><span class=o>$</span><span class=n>Pos</span><span class=p>,</span><span class=n>finalmerge</span><span class=o>$</span><span class=n>DR2_nonfiltered_BG5</span><span class=p>)</span>
<a id=__codelineno-43-3 name=__codelineno-43-3 href=#__codelineno-43-3></a><span class=nf>plot</span><span class=p>(</span><span class=n>finalmerge</span><span class=o>$</span><span class=n>Pos</span><span class=p>,</span><span class=n>finalmerge</span><span class=o>$</span><span class=n>Rsq_filtered_minimac3</span><span class=p>)</span>
<a id=__codelineno-43-4 name=__codelineno-43-4 href=#__codelineno-43-4></a><span class=nf>plot</span><span class=p>(</span><span class=n>finalmerge</span><span class=o>$</span><span class=n>Pos</span><span class=p>,</span><span class=n>finalmerge</span><span class=o>$</span><span class=n>Rsq_nonfiltered_minimac3</span><span class=p>)</span>
</code></pre></div> </div> <p>Since we have large numbers of snps in this region, it will be easier to use a bin plot other than just scatter plots. <code>ggplot2</code> has a number of functions that are useful for visualising dense data, and we will use <code>geom_hex()</code>. You don't need to run this part for yourself.</p> <div class="admonition r-project"> <p class=admonition-title>code</p> <div class=highlight><pre><span></span><code><a id=__codelineno-44-1 name=__codelineno-44-1 href=#__codelineno-44-1></a><span class=nf>library</span><span class=p>(</span><span class=n>ggplot2</span><span class=p>)</span>
<a id=__codelineno-44-2 name=__codelineno-44-2 href=#__codelineno-44-2></a><span class=nf>ggplot</span><span class=p>(</span><span class=n>finalmerge</span><span class=p>,</span><span class=nf>aes</span><span class=p>(</span><span class=n>x</span><span class=o>=</span><span class=n>Pos</span><span class=p>,</span><span class=n>y</span><span class=o>=</span><span class=n>DR2_filtered_BG5</span><span class=p>))</span><span class=w> </span><span class=o>+</span><span class=w> </span><span class=nf>geom_hex</span><span class=p>()</span>
<a id=__codelineno-44-3 name=__codelineno-44-3 href=#__codelineno-44-3></a><span class=nf>ggplot</span><span class=p>(</span><span class=n>finalmerge</span><span class=p>,</span><span class=nf>aes</span><span class=p>(</span><span class=n>x</span><span class=o>=</span><span class=n>Pos</span><span class=p>,</span><span class=n>y</span><span class=o>=</span><span class=n>DR2_nonfiltered_BG5</span><span class=p>))</span><span class=w> </span><span class=o>+</span><span class=w> </span><span class=nf>geom_hex</span><span class=p>()</span>
<a id=__codelineno-44-4 name=__codelineno-44-4 href=#__codelineno-44-4></a><span class=nf>ggplot</span><span class=p>(</span><span class=n>finalmerge</span><span class=p>,</span><span class=nf>aes</span><span class=p>(</span><span class=n>x</span><span class=o>=</span><span class=n>Pos</span><span class=p>,</span><span class=n>y</span><span class=o>=</span><span class=n>Rsq_filtered_minimac3</span><span class=p>))</span><span class=w> </span><span class=o>+</span><span class=w> </span><span class=nf>geom_hex</span><span class=p>()</span>
<a id=__codelineno-44-5 name=__codelineno-44-5 href=#__codelineno-44-5></a><span class=nf>ggplot</span><span class=p>(</span><span class=n>finalmerge</span><span class=p>,</span><span class=nf>aes</span><span class=p>(</span><span class=n>x</span><span class=o>=</span><span class=n>Pos</span><span class=p>,</span><span class=n>y</span><span class=o>=</span><span class=n>Rsq_nonfiltered_minimac3</span><span class=p>))</span><span class=w> </span><span class=o>+</span><span class=w> </span><span class=nf>geom_hex</span><span class=p>()</span>
</code></pre></div> </div> <p>Additional Plots for interest looking at Imputation Accuracy</p> <div class="admonition r-project"> <p class=admonition-title>code</p> <div class=highlight><pre><span></span><code><a id=__codelineno-45-1 name=__codelineno-45-1 href=#__codelineno-45-1></a><span class=nf>ggplot</span><span class=p>(</span><span class=n>finalmerge</span><span class=p>,</span>
<a id=__codelineno-45-2 name=__codelineno-45-2 href=#__codelineno-45-2></a><span class=w>   </span><span class=nf>aes</span><span class=p>(</span><span class=n>x</span><span class=o>=</span><span class=n>DR2_filtered_BG5</span><span class=p>,</span><span class=n>y</span><span class=o>=</span><span class=n>Rsq_filtered_minimac3</span><span class=p>,</span><span class=w> </span><span class=n>colour</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>Pos</span><span class=p>))</span><span class=w> </span><span class=o>+</span>
<a id=__codelineno-45-3 name=__codelineno-45-3 href=#__codelineno-45-3></a><span class=w>  </span><span class=nf>geom_jitter</span><span class=p>(</span><span class=n>alpha</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>0.2</span><span class=p>,</span><span class=w> </span><span class=n>width</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>0.01</span><span class=p>)</span><span class=w> </span><span class=o>+</span>
<a id=__codelineno-45-4 name=__codelineno-45-4 href=#__codelineno-45-4></a><span class=w>  </span><span class=nf>labs</span><span class=p>(</span><span class=n>x</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>expression</span><span class=p>(</span><span class=nf>paste</span><span class=p>(</span><span class=s>&quot;Beagle5 (&quot;</span><span class=p>,</span><span class=n>R</span><span class=o>^</span><span class=p>{</span><span class=m>2</span><span class=p>},</span><span class=s>&quot;)&quot;</span><span class=p>)),</span>
<a id=__codelineno-45-5 name=__codelineno-45-5 href=#__codelineno-45-5></a><span class=w>   </span><span class=n>y</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>expression</span><span class=p>(</span><span class=nf>paste</span><span class=p>(</span><span class=s>&quot;Minimac3 (&quot;</span><span class=p>,</span><span class=n>R</span><span class=o>^</span><span class=p>{</span><span class=m>2</span><span class=p>},</span><span class=s>&quot;)&quot;</span><span class=p>)),</span>
<a id=__codelineno-45-6 name=__codelineno-45-6 href=#__codelineno-45-6></a><span class=w>   </span><span class=n>title</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;Genotype Accuracy Comparison&quot;</span><span class=p>,</span>
<a id=__codelineno-45-7 name=__codelineno-45-7 href=#__codelineno-45-7></a><span class=w>   </span><span class=n>subtitle</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>expression</span><span class=p>(</span><span class=nf>paste</span><span class=p>(</span><span class=n>R</span><span class=o>^</span><span class=p>{</span><span class=m>2</span><span class=p>},</span><span class=s>&quot; between Minimac3 and Beagle5&quot;</span><span class=p>))</span>
<a id=__codelineno-45-8 name=__codelineno-45-8 href=#__codelineno-45-8></a><span class=w>  </span><span class=p>)</span><span class=w> </span><span class=o>+</span>
<a id=__codelineno-45-9 name=__codelineno-45-9 href=#__codelineno-45-9></a><span class=w>  </span><span class=nf>theme_bw</span><span class=p>()</span>
<a id=__codelineno-45-10 name=__codelineno-45-10 href=#__codelineno-45-10></a>
<a id=__codelineno-45-11 name=__codelineno-45-11 href=#__codelineno-45-11></a><span class=c1># Venn Diagram</span>
<a id=__codelineno-45-12 name=__codelineno-45-12 href=#__codelineno-45-12></a><span class=c1>#install.packages(&quot;ggvenn&quot;)</span>
<a id=__codelineno-45-13 name=__codelineno-45-13 href=#__codelineno-45-13></a><span class=nf>library</span><span class=p>(</span><span class=s>&quot;ggvenn&quot;</span><span class=p>)</span>
<a id=__codelineno-45-14 name=__codelineno-45-14 href=#__codelineno-45-14></a>
<a id=__codelineno-45-15 name=__codelineno-45-15 href=#__codelineno-45-15></a><span class=n>list_venn</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>list</span><span class=p>(</span><span class=n>Beagle5</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>which</span><span class=p>(</span><span class=n>finalmerge</span><span class=o>$</span><span class=n>DR2_filtered_BG5</span><span class=w> </span><span class=o>&gt;=</span><span class=w> </span><span class=m>0.9</span><span class=p>),</span>
<a id=__codelineno-45-16 name=__codelineno-45-16 href=#__codelineno-45-16></a><span class=w>              </span><span class=n>Minimac3</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>which</span><span class=p>(</span><span class=n>finalmerge</span><span class=o>$</span><span class=n>Rsq_filtered_minimac3</span><span class=w> </span><span class=o>&gt;=</span><span class=w> </span><span class=m>0.9</span><span class=p>))</span>
<a id=__codelineno-45-17 name=__codelineno-45-17 href=#__codelineno-45-17></a><span class=nf>ggvenn</span><span class=p>(</span><span class=n>list_venn</span><span class=p>,</span><span class=w> </span><span class=nf>c</span><span class=p>(</span><span class=s>&quot;Beagle5&quot;</span><span class=p>,</span><span class=w> </span><span class=s>&quot;Minimac3&quot;</span><span class=p>))</span>
</code></pre></div> </div> </article> </div> <script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script> </div> <button type=button class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg> Back to top </button> </main> <footer class=md-footer> <nav class="md-footer__inner md-grid" aria-label=Footer> <a href=.. class="md-footer__link md-footer__link--prev" aria-label="Previous: Home"> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </div> <div class=md-footer__title> <span class=md-footer__direction> Previous </span> <div class=md-ellipsis> Home </div> </div> </a> <a href=../supplementary_1/ class="md-footer__link md-footer__link--next" aria-label="Next: Supp 1: Imputation Introduction"> <div class=md-footer__title> <span class=md-footer__direction> Next </span> <div class=md-ellipsis> Supp 1: Imputation Introduction </div> </div> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg> </div> </a> </nav> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> <div class=md-copyright__highlight> Imputation workshop material is licensed under a <a rel=noopener target=_blank href=https://www.apache.org/licenses/LICENSE-2.0>APACHE LICENSE, VERSION 2.0</a> </div> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"base": "..", "features": ["navigation.tracking", "navigation.sections", "navigation.expand", "navigation.top", "navigation.footer", "content.code.copy", "content.code.annotate"], "search": "../assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script> <script src=../assets/javascripts/bundle.ad660dcc.min.js></script> <script src=../javascripts/mathjax.js></script> <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script> <script src=https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js></script> <script id=init-glightbox>const lightbox = GLightbox({"touchNavigation": true, "loop": false, "zoomable": true, "draggable": true, "openEffect": "zoom", "closeEffect": "zoom", "slideEffect": "slide"});
document$.subscribe(() => { lightbox.reload() });
</script></body> </html>